{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess\n",
    "import tempfile\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "train_filepath = \"train_workspace/\"\n",
    "animalList = [\"Human\",\"Pig\",\"Chicken\",\"Rat\",\"Mouse\",\"Dog\"]\n",
    "levelList = [\"Strict\",\"Relaxed\",\"Intermediate\"]\n",
    "dataList = [\"Ohnologs\",\"No-Ohnologs\",\"Paralogs\"]\n",
    "dataNameList = [\"ohnologs\",\"no-ohnologs\",\"paralog\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "working_level = \"Intermediate\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_df_dict(level):\n",
    "    df_dict = {}    \n",
    "    current_file_path = train_filepath + level + \"/\"\n",
    "    for (dataName,data) in zip(dataNameList,dataList):                \n",
    "        df_dict[data] = pd.read_pickle(current_file_path + dataName + \"-complete_v2.pkl\")        \n",
    "    return df_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Is_Ohnolog</th>\n",
       "      <th>Is_Paralog</th>\n",
       "      <th>Sequence-1</th>\n",
       "      <th>Sequence-1 GC</th>\n",
       "      <th>Sequence-1 Id</th>\n",
       "      <th>Sequence-1 Length</th>\n",
       "      <th>Sequence-1-Transcript Id</th>\n",
       "      <th>Sequence-2</th>\n",
       "      <th>Sequence-2 GC</th>\n",
       "      <th>Sequence-2 Id</th>\n",
       "      <th>Sequence-2 Length</th>\n",
       "      <th>Sequence-2-Transcript Id</th>\n",
       "      <th>Species</th>\n",
       "      <th>Sequence-1-Transcript-Version</th>\n",
       "      <th>Sequence-2-Transcript-Version</th>\n",
       "      <th>Sequence-1-Transcript-Biotype</th>\n",
       "      <th>Sequence-2-Transcript-Biotype</th>\n",
       "      <th>Sequence-1-Chromosome</th>\n",
       "      <th>Sequence-2-Chromosome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>CTTTGGAAGTCCTATGAGGGACCATTTACGGTTTCCTCAGTAATTT...</td>\n",
       "      <td>0.429090</td>\n",
       "      <td>ENSG00000095464</td>\n",
       "      <td>3307</td>\n",
       "      <td>ENST00000371447</td>\n",
       "      <td>AGTATGTTTTGCAGACAAGACCCAGAGAAGTCCAGACTGGACTTGT...</td>\n",
       "      <td>0.469856</td>\n",
       "      <td>ENSG00000132915</td>\n",
       "      <td>5706</td>\n",
       "      <td>ENST00000508173</td>\n",
       "      <td>Human</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>CGTTTTGGCAAGGGATTAAAGTGCTCCCCCCTGTGGCAGCAGTGAC...</td>\n",
       "      <td>0.441268</td>\n",
       "      <td>ENSG00000077684</td>\n",
       "      <td>5772</td>\n",
       "      <td>ENST00000226319</td>\n",
       "      <td>ATACAATAGTGCTCCGCGCCGCCTCAGCCGCCGCCGCCGCCCAACC...</td>\n",
       "      <td>0.450953</td>\n",
       "      <td>ENSG00000102221</td>\n",
       "      <td>4934</td>\n",
       "      <td>ENST00000614628</td>\n",
       "      <td>Human</td>\n",
       "      <td>10</td>\n",
       "      <td>4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>4</td>\n",
       "      <td>X</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>AGTCAACCTCTGGAAGTAAGTCAACTCCATTCTGAAAAAGAAGAGT...</td>\n",
       "      <td>0.357972</td>\n",
       "      <td>ENSG00000109158</td>\n",
       "      <td>11973</td>\n",
       "      <td>ENST00000264318</td>\n",
       "      <td>ACATAATCTAAGACCACAAACCACCTTGTTCCACGTGAGAAGGAAA...</td>\n",
       "      <td>0.402006</td>\n",
       "      <td>ENSG00000145863</td>\n",
       "      <td>2393</td>\n",
       "      <td>ENST00000523217</td>\n",
       "      <td>Human</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ACCACAGAGGCGTCTGGCTAACTCATCTCCAGACCTAAGTTGGGAA...</td>\n",
       "      <td>0.572634</td>\n",
       "      <td>ENSG00000137216</td>\n",
       "      <td>3318</td>\n",
       "      <td>ENST00000259746</td>\n",
       "      <td>CAGTCTCCTGCCATGCAAAGAAGGCTCATAGAGTTGCTTTGGAAGT...</td>\n",
       "      <td>0.516923</td>\n",
       "      <td>ENSG00000196187</td>\n",
       "      <td>5407</td>\n",
       "      <td>ENST00000474478</td>\n",
       "      <td>Human</td>\n",
       "      <td>13</td>\n",
       "      <td>5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>GGGCCTCCTCCCTGGGGTGTGAGCAGGTCGGCGCGCCACACTTCTC...</td>\n",
       "      <td>0.468779</td>\n",
       "      <td>ENSG00000113396</td>\n",
       "      <td>3219</td>\n",
       "      <td>ENST00000262462</td>\n",
       "      <td>AGTCCTGCCCGGAACCCCCGGCAACGCGCATACGACTACACCTGCT...</td>\n",
       "      <td>0.489140</td>\n",
       "      <td>ENSG00000140284</td>\n",
       "      <td>2394</td>\n",
       "      <td>ENST00000267842</td>\n",
       "      <td>Human</td>\n",
       "      <td>8</td>\n",
       "      <td>9</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>5</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Is_Ohnolog  Is_Paralog                                         Sequence-1  \\\n",
       "0          1         0.0  CTTTGGAAGTCCTATGAGGGACCATTTACGGTTTCCTCAGTAATTT...   \n",
       "1          1         0.0  CGTTTTGGCAAGGGATTAAAGTGCTCCCCCCTGTGGCAGCAGTGAC...   \n",
       "2          1         0.0  AGTCAACCTCTGGAAGTAAGTCAACTCCATTCTGAAAAAGAAGAGT...   \n",
       "3          1         0.0  ACCACAGAGGCGTCTGGCTAACTCATCTCCAGACCTAAGTTGGGAA...   \n",
       "4          1         0.0  GGGCCTCCTCCCTGGGGTGTGAGCAGGTCGGCGCGCCACACTTCTC...   \n",
       "\n",
       "   Sequence-1 GC    Sequence-1 Id  Sequence-1 Length Sequence-1-Transcript Id  \\\n",
       "0       0.429090  ENSG00000095464               3307          ENST00000371447   \n",
       "1       0.441268  ENSG00000077684               5772          ENST00000226319   \n",
       "2       0.357972  ENSG00000109158              11973          ENST00000264318   \n",
       "3       0.572634  ENSG00000137216               3318          ENST00000259746   \n",
       "4       0.468779  ENSG00000113396               3219          ENST00000262462   \n",
       "\n",
       "                                          Sequence-2  Sequence-2 GC  \\\n",
       "0  AGTATGTTTTGCAGACAAGACCCAGAGAAGTCCAGACTGGACTTGT...       0.469856   \n",
       "1  ATACAATAGTGCTCCGCGCCGCCTCAGCCGCCGCCGCCGCCCAACC...       0.450953   \n",
       "2  ACATAATCTAAGACCACAAACCACCTTGTTCCACGTGAGAAGGAAA...       0.402006   \n",
       "3  CAGTCTCCTGCCATGCAAAGAAGGCTCATAGAGTTGCTTTGGAAGT...       0.516923   \n",
       "4  AGTCCTGCCCGGAACCCCCGGCAACGCGCATACGACTACACCTGCT...       0.489140   \n",
       "\n",
       "     Sequence-2 Id  Sequence-2 Length Sequence-2-Transcript Id Species  \\\n",
       "0  ENSG00000132915               5706          ENST00000508173   Human   \n",
       "1  ENSG00000102221               4934          ENST00000614628   Human   \n",
       "2  ENSG00000145863               2393          ENST00000523217   Human   \n",
       "3  ENSG00000196187               5407          ENST00000474478   Human   \n",
       "4  ENSG00000140284               2394          ENST00000267842   Human   \n",
       "\n",
       "   Sequence-1-Transcript-Version  Sequence-2-Transcript-Version  \\\n",
       "0                              3                              5   \n",
       "1                             10                              4   \n",
       "2                              3                              5   \n",
       "3                             13                              5   \n",
       "4                              8                              9   \n",
       "\n",
       "  Sequence-1-Transcript-Biotype Sequence-2-Transcript-Biotype  \\\n",
       "0                protein_coding               retained_intron   \n",
       "1                protein_coding                protein_coding   \n",
       "2                protein_coding                protein_coding   \n",
       "3                protein_coding          processed_transcript   \n",
       "4                protein_coding                protein_coding   \n",
       "\n",
       "  Sequence-1-Chromosome Sequence-2-Chromosome  \n",
       "0                    10                     5  \n",
       "1                     4                     X  \n",
       "2                     4                     5  \n",
       "3                     6                     1  \n",
       "4                     5                    15  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_dict = get_df_dict(working_level)\n",
    "df_dict[\"Ohnologs\"].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_fasta_file(gene_id,transcript_id,sequence,version,biotype,file):\n",
    "    header = \">\" + transcript_id + \".\" + str(version) + \" \"+ gene_id + \":\" + transcript_id + \".\" + str(version) + \" cdna:\" + biotype + \"\\n\"\n",
    "    file.write(header + sequence)    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "columns_name = [\"Querry_Seq_Id\",\"Subject_Seq_Id\",\"Percent Identical Matches\",\"Aligment Length\",\"Nr Mismatch\",\"Nr Gap Open\",\"Start Align\",\n",
    "               \"End Align\",\"Start Align Sub\",\"End Align Sub\",\"Evalue\",\"Bit Score\"]\n",
    "relevant_columns = [\"Percent Identical Matches\",\"Aligment Length\",\"Nr Mismatch\",\"Nr Gap Open\",\"Evalue\",\"Bit Score\"]\n",
    "\n",
    "blast_types = [\"High\",\"Medium\",\"Low\",\"Total\"]\n",
    "relevant_col_Total = [i + \"_Total\" for i in relevant_columns] \n",
    "relevant_col_High = [i + \"_High\" for i in relevant_columns]\n",
    "relevant_col_Medium = [i + \"_Medium\" for i in relevant_columns]\n",
    "relevant_col_Low = [i + \"_Low\" for i in relevant_columns]\n",
    "total_columns = (relevant_col_High + relevant_col_Medium + relevant_col_Low + relevant_col_Total)\n",
    "\n",
    "df_blast = {}\n",
    "for dataName in dataList:\n",
    "    df = df_dict[dataName]    \n",
    "    df_blast[dataName] = pd.DataFrame(columns=total_columns)\n",
    "    \n",
    "    list_dict_cols = {}\n",
    "    list_count = []\n",
    "    for colName in total_columns:\n",
    "        list_dict_cols[colName] = []\n",
    "        \n",
    "    for index, row in df.iterrows():     \n",
    "        with open(\"gene1.fa\",\"w+\") as f1,open(\"gene2.fa\",\"w+\") as f2:      \n",
    "            f1.seek(0)\n",
    "            f1.truncate()\n",
    "            f2.seek(0)\n",
    "            f2.truncate()\n",
    "            generate_fasta_file(row[\"Sequence-1 Id\"],row[\"Sequence-1-Transcript Id\"],row[\"Sequence-1\"],row[\"Sequence-1-Transcript-Version\"],row[\"Sequence-1-Transcript-Biotype\"],f1)\n",
    "            generate_fasta_file(row[\"Sequence-2 Id\"],row[\"Sequence-2-Transcript Id\"],row[\"Sequence-2\"],row[\"Sequence-2-Transcript-Version\"],row[\"Sequence-2-Transcript-Biotype\"],f2)                 \n",
    "        list_args = ['anaconda3/envs/thesis3/bin/blastn','-query','gene1.fa','-subject','gene2.fa','-word_size','11','-outfmt','10']\n",
    "        blast_querry = subprocess.Popen(list_args, \n",
    "        stdout=subprocess.PIPE, \n",
    "        stderr=subprocess.STDOUT,\n",
    "        universal_newlines=True)\n",
    "        stdout,stderr = blast_querry.communicate()      \n",
    "        df_temp = pd.DataFrame([i.split(',') for i in stdout.split('\\n') if i != ''],columns=columns_name)\n",
    "        df_temp = df_temp[relevant_columns]\n",
    "        \n",
    "        df_temp[relevant_columns] = df_temp[relevant_columns].astype(float)\n",
    "                \n",
    "        matches_high = df_temp.loc[(df_temp[\"Evalue\"]<=1e-50)]\n",
    "        matches_medium = df_temp.loc[(df_temp[\"Evalue\"]>1e-50) & (df_temp[\"Evalue\"]<0.01)]\n",
    "        matches_low = df_temp.loc[(df_temp[\"Evalue\"]>0.01) & (df_temp[\"Evalue\"]<10)]\n",
    "              \n",
    "        df_dict_level = {}\n",
    "        df_dict_level[\"High\"] = matches_high\n",
    "        df_dict_level[\"Medium\"] = matches_medium\n",
    "        df_dict_level[\"Low\"] = matches_low\n",
    "        df_dict_level[\"Total\"] = df_temp\n",
    "        \n",
    "        list_count.append(len(df_temp))\n",
    "        \n",
    "        for blast_type in blast_types:\n",
    "            for relevant_column in relevant_columns:\n",
    "                list_dict_cols[relevant_column + \"_\" + blast_type].append(df_dict_level[blast_type][relevant_column].mean())\n",
    "                \n",
    "    df_blast[dataName]=pd.DataFrame.from_dict(list_dict_cols,orient='index').transpose()\n",
    "    df_blast[dataName][\"Nr Hits\"] = list_count                    \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Percent Identical Matches_High</th>\n",
       "      <th>Aligment Length_High</th>\n",
       "      <th>Nr Mismatch_High</th>\n",
       "      <th>Nr Gap Open_High</th>\n",
       "      <th>Evalue_High</th>\n",
       "      <th>Bit Score_High</th>\n",
       "      <th>Percent Identical Matches_Medium</th>\n",
       "      <th>Aligment Length_Medium</th>\n",
       "      <th>Nr Mismatch_Medium</th>\n",
       "      <th>Nr Gap Open_Medium</th>\n",
       "      <th>...</th>\n",
       "      <th>Nr Gap Open_Low</th>\n",
       "      <th>Evalue_Low</th>\n",
       "      <th>Bit Score_Low</th>\n",
       "      <th>Percent Identical Matches_Total</th>\n",
       "      <th>Aligment Length_Total</th>\n",
       "      <th>Nr Mismatch_Total</th>\n",
       "      <th>Nr Gap Open_Total</th>\n",
       "      <th>Evalue_Total</th>\n",
       "      <th>Bit Score_Total</th>\n",
       "      <th>Nr Hits</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>73.058</td>\n",
       "      <td>823.5</td>\n",
       "      <td>194.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>1.425000e-53</td>\n",
       "      <td>272.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.828000</td>\n",
       "      <td>22.273333</td>\n",
       "      <td>96.042529</td>\n",
       "      <td>107.352941</td>\n",
       "      <td>22.941176</td>\n",
       "      <td>2.705882</td>\n",
       "      <td>4.260000</td>\n",
       "      <td>51.652941</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>78.307</td>\n",
       "      <td>567.0</td>\n",
       "      <td>117.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>5.180000e-102</td>\n",
       "      <td>361.0</td>\n",
       "      <td>85.7505</td>\n",
       "      <td>31.5</td>\n",
       "      <td>4.50</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.272727</td>\n",
       "      <td>2.195455</td>\n",
       "      <td>23.963636</td>\n",
       "      <td>95.077926</td>\n",
       "      <td>35.592593</td>\n",
       "      <td>4.925926</td>\n",
       "      <td>0.518519</td>\n",
       "      <td>5.709458</td>\n",
       "      <td>35.981481</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>77.410</td>\n",
       "      <td>726.0</td>\n",
       "      <td>154.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>6.540000e-121</td>\n",
       "      <td>424.0</td>\n",
       "      <td>76.9230</td>\n",
       "      <td>104.0</td>\n",
       "      <td>24.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>2.126667</td>\n",
       "      <td>23.900000</td>\n",
       "      <td>92.587125</td>\n",
       "      <td>114.250000</td>\n",
       "      <td>22.500000</td>\n",
       "      <td>1.250000</td>\n",
       "      <td>1.595000</td>\n",
       "      <td>78.450000</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>86.0485</td>\n",
       "      <td>45.5</td>\n",
       "      <td>6.50</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.424444</td>\n",
       "      <td>22.344444</td>\n",
       "      <td>97.938150</td>\n",
       "      <td>15.200000</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.982011</td>\n",
       "      <td>25.025000</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>79.3405</td>\n",
       "      <td>90.0</td>\n",
       "      <td>19.25</td>\n",
       "      <td>1.25</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.409714</td>\n",
       "      <td>23.271429</td>\n",
       "      <td>90.491909</td>\n",
       "      <td>42.818182</td>\n",
       "      <td>7.818182</td>\n",
       "      <td>0.454545</td>\n",
       "      <td>0.897455</td>\n",
       "      <td>33.854545</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Percent Identical Matches_High  Aligment Length_High  Nr Mismatch_High  \\\n",
       "0                          73.058                 823.5             194.0   \n",
       "1                          78.307                 567.0             117.0   \n",
       "2                          77.410                 726.0             154.0   \n",
       "3                           0.000                   0.0               0.0   \n",
       "4                           0.000                   0.0               0.0   \n",
       "\n",
       "   Nr Gap Open_High    Evalue_High  Bit Score_High  \\\n",
       "0              23.0   1.425000e-53           272.0   \n",
       "1               6.0  5.180000e-102           361.0   \n",
       "2               9.0  6.540000e-121           424.0   \n",
       "3               0.0   0.000000e+00             0.0   \n",
       "4               0.0   0.000000e+00             0.0   \n",
       "\n",
       "   Percent Identical Matches_Medium  Aligment Length_Medium  \\\n",
       "0                            0.0000                     0.0   \n",
       "1                           85.7505                    31.5   \n",
       "2                           76.9230                   104.0   \n",
       "3                           86.0485                    45.5   \n",
       "4                           79.3405                    90.0   \n",
       "\n",
       "   Nr Mismatch_Medium  Nr Gap Open_Medium  ...  Nr Gap Open_Low  Evalue_Low  \\\n",
       "0                0.00                0.00  ...         0.000000    4.828000   \n",
       "1                4.50                0.00  ...         0.272727    2.195455   \n",
       "2               24.00                0.00  ...         0.166667    2.126667   \n",
       "3                6.50                0.00  ...         0.000000    4.424444   \n",
       "4               19.25                1.25  ...         0.000000    1.409714   \n",
       "\n",
       "   Bit Score_Low  Percent Identical Matches_Total  Aligment Length_Total  \\\n",
       "0      22.273333                        96.042529             107.352941   \n",
       "1      23.963636                        95.077926              35.592593   \n",
       "2      23.900000                        92.587125             114.250000   \n",
       "3      22.344444                        97.938150              15.200000   \n",
       "4      23.271429                        90.491909              42.818182   \n",
       "\n",
       "   Nr Mismatch_Total  Nr Gap Open_Total  Evalue_Total  Bit Score_Total  \\\n",
       "0          22.941176           2.705882      4.260000        51.652941   \n",
       "1           4.925926           0.518519      5.709458        35.981481   \n",
       "2          22.500000           1.250000      1.595000        78.450000   \n",
       "3           0.750000           0.000000      3.982011        25.025000   \n",
       "4           7.818182           0.454545      0.897455        33.854545   \n",
       "\n",
       "   Nr Hits  \n",
       "0       17  \n",
       "1       27  \n",
       "2        8  \n",
       "3       20  \n",
       "4       11  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_ohnologs_no_nan  = df_blast[\"Ohnologs\"].fillna(0)\n",
    "df_ohnologs_no_nan.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Percent Identical Matches_High</th>\n",
       "      <th>Aligment Length_High</th>\n",
       "      <th>Nr Mismatch_High</th>\n",
       "      <th>Nr Gap Open_High</th>\n",
       "      <th>Evalue_High</th>\n",
       "      <th>Bit Score_High</th>\n",
       "      <th>Percent Identical Matches_Medium</th>\n",
       "      <th>Aligment Length_Medium</th>\n",
       "      <th>Nr Mismatch_Medium</th>\n",
       "      <th>Nr Gap Open_Medium</th>\n",
       "      <th>...</th>\n",
       "      <th>Nr Gap Open_Low</th>\n",
       "      <th>Evalue_Low</th>\n",
       "      <th>Bit Score_Low</th>\n",
       "      <th>Percent Identical Matches_Total</th>\n",
       "      <th>Aligment Length_Total</th>\n",
       "      <th>Nr Mismatch_Total</th>\n",
       "      <th>Nr Gap Open_Total</th>\n",
       "      <th>Evalue_Total</th>\n",
       "      <th>Bit Score_Total</th>\n",
       "      <th>Nr Hits</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.662857</td>\n",
       "      <td>22.728571</td>\n",
       "      <td>98.979571</td>\n",
       "      <td>12.142857</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.662857</td>\n",
       "      <td>22.728571</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.988250</td>\n",
       "      <td>22.918750</td>\n",
       "      <td>98.314375</td>\n",
       "      <td>12.750000</td>\n",
       "      <td>0.312500</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.988250</td>\n",
       "      <td>22.918750</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.200000</td>\n",
       "      <td>21.400000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>11.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.200000</td>\n",
       "      <td>21.400000</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.416667</td>\n",
       "      <td>2.104833</td>\n",
       "      <td>23.575000</td>\n",
       "      <td>97.439083</td>\n",
       "      <td>14.000000</td>\n",
       "      <td>0.083333</td>\n",
       "      <td>0.416667</td>\n",
       "      <td>2.104833</td>\n",
       "      <td>23.575000</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Percent Identical Matches_High  Aligment Length_High  Nr Mismatch_High  \\\n",
       "0                             0.0                   0.0               0.0   \n",
       "1                             0.0                   0.0               0.0   \n",
       "2                             0.0                   0.0               0.0   \n",
       "3                             0.0                   0.0               0.0   \n",
       "4                             0.0                   0.0               0.0   \n",
       "\n",
       "   Nr Gap Open_High  Evalue_High  Bit Score_High  \\\n",
       "0               0.0          0.0             0.0   \n",
       "1               0.0          0.0             0.0   \n",
       "2               0.0          0.0             0.0   \n",
       "3               0.0          0.0             0.0   \n",
       "4               0.0          0.0             0.0   \n",
       "\n",
       "   Percent Identical Matches_Medium  Aligment Length_Medium  \\\n",
       "0                               0.0                     0.0   \n",
       "1                               0.0                     0.0   \n",
       "2                               0.0                     0.0   \n",
       "3                               0.0                     0.0   \n",
       "4                               0.0                     0.0   \n",
       "\n",
       "   Nr Mismatch_Medium  Nr Gap Open_Medium  ...  Nr Gap Open_Low  Evalue_Low  \\\n",
       "0                 0.0                 0.0  ...         0.000000    2.662857   \n",
       "1                 0.0                 0.0  ...         0.000000    0.000000   \n",
       "2                 0.0                 0.0  ...         0.000000    1.988250   \n",
       "3                 0.0                 0.0  ...         0.000000    3.200000   \n",
       "4                 0.0                 0.0  ...         0.416667    2.104833   \n",
       "\n",
       "   Bit Score_Low  Percent Identical Matches_Total  Aligment Length_Total  \\\n",
       "0      22.728571                        98.979571              12.142857   \n",
       "1       0.000000                         0.000000               0.000000   \n",
       "2      22.918750                        98.314375              12.750000   \n",
       "3      21.400000                       100.000000              11.000000   \n",
       "4      23.575000                        97.439083              14.000000   \n",
       "\n",
       "   Nr Mismatch_Total  Nr Gap Open_Total  Evalue_Total  Bit Score_Total  \\\n",
       "0           0.142857           0.000000      2.662857        22.728571   \n",
       "1           0.000000           0.000000      0.000000         0.000000   \n",
       "2           0.312500           0.000000      1.988250        22.918750   \n",
       "3           0.000000           0.000000      3.200000        21.400000   \n",
       "4           0.083333           0.416667      2.104833        23.575000   \n",
       "\n",
       "   Nr Hits  \n",
       "0        7  \n",
       "1        0  \n",
       "2       16  \n",
       "3        6  \n",
       "4       12  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_no_ohnologs_no_nan = df_blast[\"No-Ohnologs\"].fillna(0)\n",
    "df_no_ohnologs_no_nan.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Percent Identical Matches_High</th>\n",
       "      <th>Aligment Length_High</th>\n",
       "      <th>Nr Mismatch_High</th>\n",
       "      <th>Nr Gap Open_High</th>\n",
       "      <th>Evalue_High</th>\n",
       "      <th>Bit Score_High</th>\n",
       "      <th>Percent Identical Matches_Medium</th>\n",
       "      <th>Aligment Length_Medium</th>\n",
       "      <th>Nr Mismatch_Medium</th>\n",
       "      <th>Nr Gap Open_Medium</th>\n",
       "      <th>...</th>\n",
       "      <th>Nr Gap Open_Low</th>\n",
       "      <th>Evalue_Low</th>\n",
       "      <th>Bit Score_Low</th>\n",
       "      <th>Percent Identical Matches_Total</th>\n",
       "      <th>Aligment Length_Total</th>\n",
       "      <th>Nr Mismatch_Total</th>\n",
       "      <th>Nr Gap Open_Total</th>\n",
       "      <th>Evalue_Total</th>\n",
       "      <th>Bit Score_Total</th>\n",
       "      <th>Nr Hits</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>82.157</td>\n",
       "      <td>48.0</td>\n",
       "      <td>7.5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.700000</td>\n",
       "      <td>22.350000</td>\n",
       "      <td>91.089333</td>\n",
       "      <td>25.333333</td>\n",
       "      <td>2.833333</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.466690</td>\n",
       "      <td>28.816667</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.272727</td>\n",
       "      <td>5.463636</td>\n",
       "      <td>23.627273</td>\n",
       "      <td>96.667091</td>\n",
       "      <td>14.363636</td>\n",
       "      <td>0.272727</td>\n",
       "      <td>0.272727</td>\n",
       "      <td>5.463636</td>\n",
       "      <td>23.627273</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>76.471</td>\n",
       "      <td>102.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>1.722300</td>\n",
       "      <td>22.900000</td>\n",
       "      <td>96.075273</td>\n",
       "      <td>21.000000</td>\n",
       "      <td>2.090909</td>\n",
       "      <td>0.363636</td>\n",
       "      <td>1.565727</td>\n",
       "      <td>25.790909</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.272727</td>\n",
       "      <td>3.754545</td>\n",
       "      <td>22.254545</td>\n",
       "      <td>97.296182</td>\n",
       "      <td>13.090909</td>\n",
       "      <td>0.181818</td>\n",
       "      <td>0.272727</td>\n",
       "      <td>3.754545</td>\n",
       "      <td>22.254545</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>88.889</td>\n",
       "      <td>27.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>5.085250</td>\n",
       "      <td>22.615000</td>\n",
       "      <td>96.371476</td>\n",
       "      <td>16.619048</td>\n",
       "      <td>1.428571</td>\n",
       "      <td>0.095238</td>\n",
       "      <td>4.843143</td>\n",
       "      <td>23.176190</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Percent Identical Matches_High  Aligment Length_High  Nr Mismatch_High  \\\n",
       "0                             0.0                   0.0               0.0   \n",
       "1                             0.0                   0.0               0.0   \n",
       "2                             0.0                   0.0               0.0   \n",
       "3                             0.0                   0.0               0.0   \n",
       "4                             0.0                   0.0               0.0   \n",
       "\n",
       "   Nr Gap Open_High  Evalue_High  Bit Score_High  \\\n",
       "0               0.0          0.0             0.0   \n",
       "1               0.0          0.0             0.0   \n",
       "2               0.0          0.0             0.0   \n",
       "3               0.0          0.0             0.0   \n",
       "4               0.0          0.0             0.0   \n",
       "\n",
       "   Percent Identical Matches_Medium  Aligment Length_Medium  \\\n",
       "0                            82.157                    48.0   \n",
       "1                             0.000                     0.0   \n",
       "2                            76.471                   102.0   \n",
       "3                             0.000                     0.0   \n",
       "4                            88.889                    27.0   \n",
       "\n",
       "   Nr Mismatch_Medium  Nr Gap Open_Medium  ...  Nr Gap Open_Low  Evalue_Low  \\\n",
       "0                 7.5                 1.0  ...         0.250000    0.700000   \n",
       "1                 0.0                 0.0  ...         0.272727    5.463636   \n",
       "2                22.0                 2.0  ...         0.200000    1.722300   \n",
       "3                 0.0                 0.0  ...         0.272727    3.754545   \n",
       "4                 3.0                 0.0  ...         0.100000    5.085250   \n",
       "\n",
       "   Bit Score_Low  Percent Identical Matches_Total  Aligment Length_Total  \\\n",
       "0      22.350000                        91.089333              25.333333   \n",
       "1      23.627273                        96.667091              14.363636   \n",
       "2      22.900000                        96.075273              21.000000   \n",
       "3      22.254545                        97.296182              13.090909   \n",
       "4      22.615000                        96.371476              16.619048   \n",
       "\n",
       "   Nr Mismatch_Total  Nr Gap Open_Total  Evalue_Total  Bit Score_Total  \\\n",
       "0           2.833333           0.500000      0.466690        28.816667   \n",
       "1           0.272727           0.272727      5.463636        23.627273   \n",
       "2           2.090909           0.363636      1.565727        25.790909   \n",
       "3           0.181818           0.272727      3.754545        22.254545   \n",
       "4           1.428571           0.095238      4.843143        23.176190   \n",
       "\n",
       "   Nr Hits  \n",
       "0        6  \n",
       "1       11  \n",
       "2       11  \n",
       "3       11  \n",
       "4       21  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_paralogs_no_nan = df_blast[\"Paralogs\"].fillna(0)\n",
    "df_paralogs_no_nan.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Percent Identical Matches_High</th>\n",
       "      <th>Aligment Length_High</th>\n",
       "      <th>Nr Mismatch_High</th>\n",
       "      <th>Nr Gap Open_High</th>\n",
       "      <th>Evalue_High</th>\n",
       "      <th>Bit Score_High</th>\n",
       "      <th>Percent Identical Matches_Medium</th>\n",
       "      <th>Aligment Length_Medium</th>\n",
       "      <th>Nr Mismatch_Medium</th>\n",
       "      <th>Nr Gap Open_Medium</th>\n",
       "      <th>...</th>\n",
       "      <th>Nr Gap Open_Low</th>\n",
       "      <th>Evalue_Low</th>\n",
       "      <th>Bit Score_Low</th>\n",
       "      <th>Percent Identical Matches_Total</th>\n",
       "      <th>Aligment Length_Total</th>\n",
       "      <th>Nr Mismatch_Total</th>\n",
       "      <th>Nr Gap Open_Total</th>\n",
       "      <th>Evalue_Total</th>\n",
       "      <th>Bit Score_Total</th>\n",
       "      <th>Nr Hits</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>22966</th>\n",
       "      <td>83.019</td>\n",
       "      <td>212.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>9.440000e-51</td>\n",
       "      <td>191.0</td>\n",
       "      <td>85.039667</td>\n",
       "      <td>53.333333</td>\n",
       "      <td>9.0</td>\n",
       "      <td>1.222222</td>\n",
       "      <td>...</td>\n",
       "      <td>0.128571</td>\n",
       "      <td>2.190800</td>\n",
       "      <td>24.791429</td>\n",
       "      <td>94.930687</td>\n",
       "      <td>21.750000</td>\n",
       "      <td>1.925000</td>\n",
       "      <td>0.275000</td>\n",
       "      <td>1.917138</td>\n",
       "      <td>28.523750</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22967</th>\n",
       "      <td>78.030</td>\n",
       "      <td>396.0</td>\n",
       "      <td>85.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>8.890000e-69</td>\n",
       "      <td>248.0</td>\n",
       "      <td>73.984000</td>\n",
       "      <td>369.000000</td>\n",
       "      <td>83.0</td>\n",
       "      <td>11.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.590000</td>\n",
       "      <td>23.300000</td>\n",
       "      <td>89.780000</td>\n",
       "      <td>136.500000</td>\n",
       "      <td>28.333333</td>\n",
       "      <td>2.166667</td>\n",
       "      <td>0.393333</td>\n",
       "      <td>79.700000</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22968</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.079000</td>\n",
       "      <td>21.400000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>11.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.079000</td>\n",
       "      <td>21.400000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22969</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.260870</td>\n",
       "      <td>3.646522</td>\n",
       "      <td>22.869565</td>\n",
       "      <td>97.518304</td>\n",
       "      <td>13.826087</td>\n",
       "      <td>0.260870</td>\n",
       "      <td>0.260870</td>\n",
       "      <td>3.646522</td>\n",
       "      <td>22.869565</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22970</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>4.832000</td>\n",
       "      <td>22.520000</td>\n",
       "      <td>96.644800</td>\n",
       "      <td>13.600000</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>4.832000</td>\n",
       "      <td>22.520000</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Percent Identical Matches_High  Aligment Length_High  Nr Mismatch_High  \\\n",
       "22966                          83.019                 212.0              34.0   \n",
       "22967                          78.030                 396.0              85.0   \n",
       "22968                           0.000                   0.0               0.0   \n",
       "22969                           0.000                   0.0               0.0   \n",
       "22970                           0.000                   0.0               0.0   \n",
       "\n",
       "       Nr Gap Open_High   Evalue_High  Bit Score_High  \\\n",
       "22966               2.0  9.440000e-51           191.0   \n",
       "22967               2.0  8.890000e-69           248.0   \n",
       "22968               0.0  0.000000e+00             0.0   \n",
       "22969               0.0  0.000000e+00             0.0   \n",
       "22970               0.0  0.000000e+00             0.0   \n",
       "\n",
       "       Percent Identical Matches_Medium  Aligment Length_Medium  \\\n",
       "22966                         85.039667               53.333333   \n",
       "22967                         73.984000              369.000000   \n",
       "22968                          0.000000                0.000000   \n",
       "22969                          0.000000                0.000000   \n",
       "22970                          0.000000                0.000000   \n",
       "\n",
       "       Nr Mismatch_Medium  Nr Gap Open_Medium  ...  Nr Gap Open_Low  \\\n",
       "22966                 9.0            1.222222  ...         0.128571   \n",
       "22967                83.0           11.000000  ...         0.000000   \n",
       "22968                 0.0            0.000000  ...         0.000000   \n",
       "22969                 0.0            0.000000  ...         0.260870   \n",
       "22970                 0.0            0.000000  ...         0.200000   \n",
       "\n",
       "       Evalue_Low  Bit Score_Low  Percent Identical Matches_Total  \\\n",
       "22966    2.190800      24.791429                        94.930687   \n",
       "22967    0.590000      23.300000                        89.780000   \n",
       "22968    0.079000      21.400000                       100.000000   \n",
       "22969    3.646522      22.869565                        97.518304   \n",
       "22970    4.832000      22.520000                        96.644800   \n",
       "\n",
       "       Aligment Length_Total  Nr Mismatch_Total  Nr Gap Open_Total  \\\n",
       "22966              21.750000           1.925000           0.275000   \n",
       "22967             136.500000          28.333333           2.166667   \n",
       "22968              11.000000           0.000000           0.000000   \n",
       "22969              13.826087           0.260870           0.260870   \n",
       "22970              13.600000           0.400000           0.200000   \n",
       "\n",
       "       Evalue_Total  Bit Score_Total  Nr Hits  \n",
       "22966      1.917138        28.523750       80  \n",
       "22967      0.393333        79.700000        6  \n",
       "22968      0.079000        21.400000        1  \n",
       "22969      3.646522        22.869565       23  \n",
       "22970      4.832000        22.520000        5  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_paralogs_no_nan.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Is_Ohnolog</th>\n",
       "      <th>Is_Paralog</th>\n",
       "      <th>Sequence-1</th>\n",
       "      <th>Sequence-1 GC</th>\n",
       "      <th>Sequence-1 Id</th>\n",
       "      <th>Sequence-1 Length</th>\n",
       "      <th>Sequence-1-Transcript Id</th>\n",
       "      <th>Sequence-2</th>\n",
       "      <th>Sequence-2 GC</th>\n",
       "      <th>Sequence-2 Id</th>\n",
       "      <th>Sequence-2 Length</th>\n",
       "      <th>Sequence-2-Transcript Id</th>\n",
       "      <th>Species</th>\n",
       "      <th>Sequence-1-Transcript-Version</th>\n",
       "      <th>Sequence-2-Transcript-Version</th>\n",
       "      <th>Sequence-1-Transcript-Biotype</th>\n",
       "      <th>Sequence-2-Transcript-Biotype</th>\n",
       "      <th>Sequence-1-Chromosome</th>\n",
       "      <th>Sequence-2-Chromosome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>499</th>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>GTCGTCCGCAAAGCCTGAGTCCTGTCCTTTCTCTCTCCCCGGACAG...</td>\n",
       "      <td>0.556992</td>\n",
       "      <td>ENSG00000111057</td>\n",
       "      <td>2088</td>\n",
       "      <td>ENST00000549078</td>\n",
       "      <td>ATGACCTCCTTCTACAGCACCTCCTCATGCCCTCTGGGTTGCACCA...</td>\n",
       "      <td>0.588475</td>\n",
       "      <td>ENSG00000108417</td>\n",
       "      <td>1475</td>\n",
       "      <td>ENST00000225550</td>\n",
       "      <td>Human</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>12</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1029</th>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>AGAGCGGGCGGGGCGGACACGCTTAGCGTAAGCGCAGGGGCCAGGA...</td>\n",
       "      <td>0.418398</td>\n",
       "      <td>ENSG00000164794</td>\n",
       "      <td>6979</td>\n",
       "      <td>ENST00000524391</td>\n",
       "      <td>GGAGCCGGCGCGGCCGCGGGCCAGGCGCCGAGGTGCGCGCGGAGCG...</td>\n",
       "      <td>0.643981</td>\n",
       "      <td>ENSG00000075043</td>\n",
       "      <td>9213</td>\n",
       "      <td>ENST00000626839</td>\n",
       "      <td>Human</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>8</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3517</th>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>AGCGGGCGGCGAGCAGCGCGGAGCGCGCGGGGCACCGAGGAGCGGG...</td>\n",
       "      <td>0.622584</td>\n",
       "      <td>ENSG00000100314</td>\n",
       "      <td>3259</td>\n",
       "      <td>ENST00000216144</td>\n",
       "      <td>CATAAGCAGCTTTATGGGTCCTGAAAGCCAAGGGTGCCCAGGGGTG...</td>\n",
       "      <td>0.548652</td>\n",
       "      <td>ENSG00000175544</td>\n",
       "      <td>2559</td>\n",
       "      <td>ENST00000325656</td>\n",
       "      <td>Human</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>22</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4305</th>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CACATGCTGGACGGATTCTGGGTCTGCTCTTTGTGCAAACAAAACA...</td>\n",
       "      <td>0.432248</td>\n",
       "      <td>ENSG00000184588</td>\n",
       "      <td>4546</td>\n",
       "      <td>ENST00000423207</td>\n",
       "      <td>TGCGTGCCTGGAGCAGCAGCGTCTCCAGGGACAGGCAGCCACCATG...</td>\n",
       "      <td>0.555183</td>\n",
       "      <td>ENSG00000133256</td>\n",
       "      <td>3280</td>\n",
       "      <td>ENST00000255622</td>\n",
       "      <td>Human</td>\n",
       "      <td>6</td>\n",
       "      <td>10</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4123</th>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CTGTTCCAGACTTCTCAGGGTTTTTACAGCACAAACAACTGCATAC...</td>\n",
       "      <td>0.432591</td>\n",
       "      <td>ENSG00000151572</td>\n",
       "      <td>3909</td>\n",
       "      <td>ENST00000392979</td>\n",
       "      <td>AGGAGCACCCGCCGCTGGCAGCCACTCTCCTGGCCGCCAAACCCCA...</td>\n",
       "      <td>0.641152</td>\n",
       "      <td>ENSG00000185101</td>\n",
       "      <td>5487</td>\n",
       "      <td>ENST00000532094</td>\n",
       "      <td>Human</td>\n",
       "      <td>7</td>\n",
       "      <td>5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>12</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Is_Ohnolog  Is_Paralog  \\\n",
       "499           0         1.0   \n",
       "1029          0         1.0   \n",
       "3517          0         1.0   \n",
       "4305          0         1.0   \n",
       "4123          0         1.0   \n",
       "\n",
       "                                             Sequence-1  Sequence-1 GC  \\\n",
       "499   GTCGTCCGCAAAGCCTGAGTCCTGTCCTTTCTCTCTCCCCGGACAG...       0.556992   \n",
       "1029  AGAGCGGGCGGGGCGGACACGCTTAGCGTAAGCGCAGGGGCCAGGA...       0.418398   \n",
       "3517  AGCGGGCGGCGAGCAGCGCGGAGCGCGCGGGGCACCGAGGAGCGGG...       0.622584   \n",
       "4305  CACATGCTGGACGGATTCTGGGTCTGCTCTTTGTGCAAACAAAACA...       0.432248   \n",
       "4123  CTGTTCCAGACTTCTCAGGGTTTTTACAGCACAAACAACTGCATAC...       0.432591   \n",
       "\n",
       "        Sequence-1 Id Sequence-1 Length Sequence-1-Transcript Id  \\\n",
       "499   ENSG00000111057              2088          ENST00000549078   \n",
       "1029  ENSG00000164794              6979          ENST00000524391   \n",
       "3517  ENSG00000100314              3259          ENST00000216144   \n",
       "4305  ENSG00000184588              4546          ENST00000423207   \n",
       "4123  ENSG00000151572              3909          ENST00000392979   \n",
       "\n",
       "                                             Sequence-2  Sequence-2 GC  \\\n",
       "499   ATGACCTCCTTCTACAGCACCTCCTCATGCCCTCTGGGTTGCACCA...       0.588475   \n",
       "1029  GGAGCCGGCGCGGCCGCGGGCCAGGCGCCGAGGTGCGCGCGGAGCG...       0.643981   \n",
       "3517  CATAAGCAGCTTTATGGGTCCTGAAAGCCAAGGGTGCCCAGGGGTG...       0.548652   \n",
       "4305  TGCGTGCCTGGAGCAGCAGCGTCTCCAGGGACAGGCAGCCACCATG...       0.555183   \n",
       "4123  AGGAGCACCCGCCGCTGGCAGCCACTCTCCTGGCCGCCAAACCCCA...       0.641152   \n",
       "\n",
       "        Sequence-2 Id Sequence-2 Length Sequence-2-Transcript Id Species  \\\n",
       "499   ENSG00000108417              1475          ENST00000225550   Human   \n",
       "1029  ENSG00000075043              9213          ENST00000626839   Human   \n",
       "3517  ENSG00000175544              2559          ENST00000325656   Human   \n",
       "4305  ENSG00000133256              3280          ENST00000255622   Human   \n",
       "4123  ENSG00000185101              5487          ENST00000532094   Human   \n",
       "\n",
       "      Sequence-1-Transcript-Version  Sequence-2-Transcript-Version  \\\n",
       "499                               5                              3   \n",
       "1029                              5                              2   \n",
       "3517                              3                              6   \n",
       "4305                              6                             10   \n",
       "4123                              7                              5   \n",
       "\n",
       "     Sequence-1-Transcript-Biotype Sequence-2-Transcript-Biotype  \\\n",
       "499                retained_intron                protein_coding   \n",
       "1029                protein_coding                protein_coding   \n",
       "3517                protein_coding                protein_coding   \n",
       "4305                protein_coding                protein_coding   \n",
       "4123                protein_coding               retained_intron   \n",
       "\n",
       "     Sequence-1-Chromosome Sequence-2-Chromosome  \n",
       "499                     12                    17  \n",
       "1029                     8                    20  \n",
       "3517                    22                    11  \n",
       "4305                     1                     4  \n",
       "4123                    12                    11  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_dict[\"Paralogs\"].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Is_Ohnolog</th>\n",
       "      <th>Is_Paralog</th>\n",
       "      <th>Sequence-1</th>\n",
       "      <th>Sequence-1 GC</th>\n",
       "      <th>Sequence-1 Id</th>\n",
       "      <th>Sequence-1 Length</th>\n",
       "      <th>Sequence-1-Transcript Id</th>\n",
       "      <th>Sequence-2</th>\n",
       "      <th>Sequence-2 GC</th>\n",
       "      <th>Sequence-2 Id</th>\n",
       "      <th>...</th>\n",
       "      <th>Nr Gap Open_Low</th>\n",
       "      <th>Evalue_Low</th>\n",
       "      <th>Bit Score_Low</th>\n",
       "      <th>Percent Identical Matches_Total</th>\n",
       "      <th>Aligment Length_Total</th>\n",
       "      <th>Nr Mismatch_Total</th>\n",
       "      <th>Nr Gap Open_Total</th>\n",
       "      <th>Evalue_Total</th>\n",
       "      <th>Bit Score_Total</th>\n",
       "      <th>Nr Hits</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>GTCGTCCGCAAAGCCTGAGTCCTGTCCTTTCTCTCTCCCCGGACAG...</td>\n",
       "      <td>0.556992</td>\n",
       "      <td>ENSG00000111057</td>\n",
       "      <td>2088</td>\n",
       "      <td>ENST00000549078</td>\n",
       "      <td>ATGACCTCCTTCTACAGCACCTCCTCATGCCCTCTGGGTTGCACCA...</td>\n",
       "      <td>0.588475</td>\n",
       "      <td>ENSG00000108417</td>\n",
       "      <td>...</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.700000</td>\n",
       "      <td>22.350000</td>\n",
       "      <td>91.089333</td>\n",
       "      <td>25.333333</td>\n",
       "      <td>2.833333</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.466690</td>\n",
       "      <td>28.816667</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>AGAGCGGGCGGGGCGGACACGCTTAGCGTAAGCGCAGGGGCCAGGA...</td>\n",
       "      <td>0.418398</td>\n",
       "      <td>ENSG00000164794</td>\n",
       "      <td>6979</td>\n",
       "      <td>ENST00000524391</td>\n",
       "      <td>GGAGCCGGCGCGGCCGCGGGCCAGGCGCCGAGGTGCGCGCGGAGCG...</td>\n",
       "      <td>0.643981</td>\n",
       "      <td>ENSG00000075043</td>\n",
       "      <td>...</td>\n",
       "      <td>0.272727</td>\n",
       "      <td>5.463636</td>\n",
       "      <td>23.627273</td>\n",
       "      <td>96.667091</td>\n",
       "      <td>14.363636</td>\n",
       "      <td>0.272727</td>\n",
       "      <td>0.272727</td>\n",
       "      <td>5.463636</td>\n",
       "      <td>23.627273</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>AGCGGGCGGCGAGCAGCGCGGAGCGCGCGGGGCACCGAGGAGCGGG...</td>\n",
       "      <td>0.622584</td>\n",
       "      <td>ENSG00000100314</td>\n",
       "      <td>3259</td>\n",
       "      <td>ENST00000216144</td>\n",
       "      <td>CATAAGCAGCTTTATGGGTCCTGAAAGCCAAGGGTGCCCAGGGGTG...</td>\n",
       "      <td>0.548652</td>\n",
       "      <td>ENSG00000175544</td>\n",
       "      <td>...</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>1.722300</td>\n",
       "      <td>22.900000</td>\n",
       "      <td>96.075273</td>\n",
       "      <td>21.000000</td>\n",
       "      <td>2.090909</td>\n",
       "      <td>0.363636</td>\n",
       "      <td>1.565727</td>\n",
       "      <td>25.790909</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CACATGCTGGACGGATTCTGGGTCTGCTCTTTGTGCAAACAAAACA...</td>\n",
       "      <td>0.432248</td>\n",
       "      <td>ENSG00000184588</td>\n",
       "      <td>4546</td>\n",
       "      <td>ENST00000423207</td>\n",
       "      <td>TGCGTGCCTGGAGCAGCAGCGTCTCCAGGGACAGGCAGCCACCATG...</td>\n",
       "      <td>0.555183</td>\n",
       "      <td>ENSG00000133256</td>\n",
       "      <td>...</td>\n",
       "      <td>0.272727</td>\n",
       "      <td>3.754545</td>\n",
       "      <td>22.254545</td>\n",
       "      <td>97.296182</td>\n",
       "      <td>13.090909</td>\n",
       "      <td>0.181818</td>\n",
       "      <td>0.272727</td>\n",
       "      <td>3.754545</td>\n",
       "      <td>22.254545</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CTGTTCCAGACTTCTCAGGGTTTTTACAGCACAAACAACTGCATAC...</td>\n",
       "      <td>0.432591</td>\n",
       "      <td>ENSG00000151572</td>\n",
       "      <td>3909</td>\n",
       "      <td>ENST00000392979</td>\n",
       "      <td>AGGAGCACCCGCCGCTGGCAGCCACTCTCCTGGCCGCCAAACCCCA...</td>\n",
       "      <td>0.641152</td>\n",
       "      <td>ENSG00000185101</td>\n",
       "      <td>...</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>5.085250</td>\n",
       "      <td>22.615000</td>\n",
       "      <td>96.371476</td>\n",
       "      <td>16.619048</td>\n",
       "      <td>1.428571</td>\n",
       "      <td>0.095238</td>\n",
       "      <td>4.843143</td>\n",
       "      <td>23.176190</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 44 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Is_Ohnolog  Is_Paralog                                         Sequence-1  \\\n",
       "0          0         1.0  GTCGTCCGCAAAGCCTGAGTCCTGTCCTTTCTCTCTCCCCGGACAG...   \n",
       "1          0         1.0  AGAGCGGGCGGGGCGGACACGCTTAGCGTAAGCGCAGGGGCCAGGA...   \n",
       "2          0         1.0  AGCGGGCGGCGAGCAGCGCGGAGCGCGCGGGGCACCGAGGAGCGGG...   \n",
       "3          0         1.0  CACATGCTGGACGGATTCTGGGTCTGCTCTTTGTGCAAACAAAACA...   \n",
       "4          0         1.0  CTGTTCCAGACTTCTCAGGGTTTTTACAGCACAAACAACTGCATAC...   \n",
       "\n",
       "   Sequence-1 GC    Sequence-1 Id Sequence-1 Length Sequence-1-Transcript Id  \\\n",
       "0       0.556992  ENSG00000111057              2088          ENST00000549078   \n",
       "1       0.418398  ENSG00000164794              6979          ENST00000524391   \n",
       "2       0.622584  ENSG00000100314              3259          ENST00000216144   \n",
       "3       0.432248  ENSG00000184588              4546          ENST00000423207   \n",
       "4       0.432591  ENSG00000151572              3909          ENST00000392979   \n",
       "\n",
       "                                          Sequence-2  Sequence-2 GC  \\\n",
       "0  ATGACCTCCTTCTACAGCACCTCCTCATGCCCTCTGGGTTGCACCA...       0.588475   \n",
       "1  GGAGCCGGCGCGGCCGCGGGCCAGGCGCCGAGGTGCGCGCGGAGCG...       0.643981   \n",
       "2  CATAAGCAGCTTTATGGGTCCTGAAAGCCAAGGGTGCCCAGGGGTG...       0.548652   \n",
       "3  TGCGTGCCTGGAGCAGCAGCGTCTCCAGGGACAGGCAGCCACCATG...       0.555183   \n",
       "4  AGGAGCACCCGCCGCTGGCAGCCACTCTCCTGGCCGCCAAACCCCA...       0.641152   \n",
       "\n",
       "     Sequence-2 Id  ... Nr Gap Open_Low Evalue_Low Bit Score_Low  \\\n",
       "0  ENSG00000108417  ...        0.250000   0.700000     22.350000   \n",
       "1  ENSG00000075043  ...        0.272727   5.463636     23.627273   \n",
       "2  ENSG00000175544  ...        0.200000   1.722300     22.900000   \n",
       "3  ENSG00000133256  ...        0.272727   3.754545     22.254545   \n",
       "4  ENSG00000185101  ...        0.100000   5.085250     22.615000   \n",
       "\n",
       "   Percent Identical Matches_Total  Aligment Length_Total Nr Mismatch_Total  \\\n",
       "0                        91.089333              25.333333          2.833333   \n",
       "1                        96.667091              14.363636          0.272727   \n",
       "2                        96.075273              21.000000          2.090909   \n",
       "3                        97.296182              13.090909          0.181818   \n",
       "4                        96.371476              16.619048          1.428571   \n",
       "\n",
       "  Nr Gap Open_Total Evalue_Total Bit Score_Total  Nr Hits  \n",
       "0          0.500000     0.466690       28.816667        6  \n",
       "1          0.272727     5.463636       23.627273       11  \n",
       "2          0.363636     1.565727       25.790909       11  \n",
       "3          0.272727     3.754545       22.254545       11  \n",
       "4          0.095238     4.843143       23.176190       21  \n",
       "\n",
       "[5 rows x 44 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "name_data_dict = {}\n",
    "name_data_dict[\"Ohnologs\"] = \"ohnologs\"\n",
    "name_data_dict[\"No-Ohnologs\"] = \"no-ohnologs\"\n",
    "name_data_dict[\"Paralogs\"] = \"paralog\"\n",
    "\n",
    "df_data_dict = {}\n",
    "df_data_dict[\"Ohnologs\"] = df_ohnologs_no_nan\n",
    "df_data_dict[\"No-Ohnologs\"] = df_paralogs_no_nan\n",
    "df_data_dict[\"Paralogs\"] = df_paralogs_no_nan\n",
    "\n",
    "for dataName in dataList:\n",
    "    df_save = pd.concat([df_dict[dataName].reset_index(drop=True),df_data_dict[dataName].reset_index(drop=True)],axis=1,sort=False)\n",
    "    df_save.to_pickle(train_filepath + working_level + \"/\" + name_data_dict[dataName] + \"-complete2.pkl\")\n",
    "    \n",
    "\n",
    "df_save.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "thesis3",
   "language": "python",
   "name": "thesis3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
