{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from random import randint\n",
    "import random\n",
    "import os\n",
    "from matplotlib import pyplot as plt\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sb\n",
    "import pickle as pkl\n",
    "import math\n",
    "random.seed(42)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "train_filepath = \"train_workspace/\"\n",
    "animalList = [\"Human\",\"Pig\",\"Chicken\",\"Rat\",\"Mouse\",\"Dog\"]\n",
    "levelList = [\"Strict\",\"Relaxed\",\"Intermediate\"]\n",
    "dataList = [\"Ohnologs\",\"No-Ohnologs\",\"Paralogs\"]\n",
    "dataNameList = [\"ohnologs\",\"no-ohnologs\",\"paralog\"]\n",
    "sb.set_style(\"darkgrid\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_df_dict(level):\n",
    "    df_dict = {}    \n",
    "    current_file_path = train_filepath + level + \"/\"\n",
    "    for (dataName,data) in zip(dataNameList,dataList):                \n",
    "        df_dict[data] = pd.read_pickle(current_file_path + dataName + \"-complete2.pkl\")        \n",
    "    return df_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mutate_simple(dna):\n",
    "    dna_list = list(dna)\n",
    "    mutation_site = random.randint(0, len(dna_list) - 1)\n",
    "    dna_list[mutation_site] = random.choice(list('ATCG'))\n",
    "    return ''.join(dna_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def draw(discrete_probdist):\n",
    "    \"\"\"\n",
    "    Draw random value from discrete probability distribution\n",
    "    represented as a dict: P(x=value) = discrete_probdist[value].\n",
    "    \"\"\"    \n",
    "    limit = 0\n",
    "    r = random.random()\n",
    "    for value in discrete_probdist:\n",
    "        limit += discrete_probdist[value]\n",
    "        if r < limit:\n",
    "            return value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Takes into account the frecuency of transitions and transversions\n",
    "def create_markov_chain():\n",
    "    markov_chain = {}\n",
    "    for from_base in 'ATGC':\n",
    "        # Generate random transition probabilities by dividing\n",
    "        # [0,1] into four intervals of random length\n",
    "        slice_points = sorted([0] + [random.random()for i in range(3)] + [1])\n",
    "        transition_probabilities = [slice_points[i+1] - slice_points[i] for i in range(4)]\n",
    "        transition_probabilities_sorted = sorted(transition_probabilities)\n",
    "        bases_string = 'ATGC'\n",
    "        if(from_base == 'A'):\n",
    "            bases_string = bases_string.replace('G','')\n",
    "            bases_string = bases_string.replace('A','')\n",
    "            bases_string = bases_string + 'A'     \n",
    "            bases_string = bases_string + 'G'            \n",
    "        if(from_base == 'G'):\n",
    "            bases_string = bases_string.replace('A','')\n",
    "            bases_string = bases_string.replace('G','')\n",
    "            bases_string = bases_string + 'G'            \n",
    "            bases_string = bases_string + 'A'            \n",
    "        if(from_base == 'C'):\n",
    "            bases_string = bases_string.replace('T','')\n",
    "            bases_string = bases_string.replace('C','')\n",
    "            bases_string = bases_string + 'C'\n",
    "            bases_string = bases_string + 'T'\n",
    "        if(from_base == 'T'):\n",
    "            bases_string = bases_string.replace('C','')\n",
    "            bases_string = bases_string.replace('T','')\n",
    "            bases_string = bases_string + 'T'   \n",
    "            bases_string = bases_string + 'C'           \n",
    "        markov_chain[from_base] = {base: p for base, p in zip(bases_string, transition_probabilities_sorted)}\n",
    "    return markov_chain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Pointwise mutation\n",
    "def mutate_via_markov_chain(dna, markov_chain,mutation_site):       \n",
    "    dna_list = list(dna)\n",
    "    from_base = dna[mutation_site]\n",
    "    if(from_base == 'N'):\n",
    "        return dna\n",
    "    to_base = draw(markov_chain[from_base])\n",
    "    dna_list[mutation_site] = to_base\n",
    "    return ''.join(dna_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_random_sequence(length,gc_percent):\n",
    "    dna_sequence = ''\n",
    "    for i in range(0,length):\n",
    "        step = random.random()\n",
    "        next_nucleotide = ''\n",
    "        if(step < gc_percent):\n",
    "            next_nucleotide =  random.choice(list('CG'))\n",
    "        else:\n",
    "            next_nucleotide =  random.choice(list('AT'))\n",
    "        dna_sequence += next_nucleotide\n",
    "    return dna_sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_gc_count(sequence):\n",
    "    count = 0\n",
    "    for c in sequence:\n",
    "        if(c == 'G'):\n",
    "            count = count + 1\n",
    "        if(c == 'C'):\n",
    "            count = count + 1            \n",
    "    return count    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def insert_str(string, str_to_insert, index):\n",
    "    return string[:index] + str_to_insert + string[index:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mutate_insertion(dna,gc_percent,insertion_length_max,mutation_site):\n",
    "    return insert_str(dna,generate_random_sequence(randint(1, insertion_length_max),gc_percent),mutation_site)        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mutate_tandem_insertion(dna,tandem_length_max,tandem_quantity_max,mutation_site):    \n",
    "    tandem_length = randint(1, tandem_length_max)\n",
    "    final_position = mutation_site + tandem_length\n",
    "    insert_string = dna[mutation_site:final_position]\n",
    "    tandem_length = randint(1, tandem_quantity_max)\n",
    "    for i in range(1,tandem_length):\n",
    "        dna = insert_str(dna,insert_string,mutation_site) \n",
    "    return dna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mutate_deletion(dna,deletion_length_max,mutation_site):\n",
    "    deletion_length =  randint(1, deletion_length_max)\n",
    "    deletion_position = mutation_site + deletion_length\n",
    "    return dna[0:mutation_site] + dna[deletion_position:]    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Likelyhood of indels are between 16% and 25%\n",
    "def generate_mutated_sequence(dna_sequence,mutation_rate):\n",
    "    mc = create_markov_chain()\n",
    "    dna_length = len(dna_sequence)\n",
    "    nr_mutations = round(dna_length * mutation_rate)  \n",
    "    for i in range(0,nr_mutations):\n",
    "        threshold = random.random()\n",
    "        mutation_site = random.randint(0, len(dna_sequence) - 1)\n",
    "        if(threshold<0.8):\n",
    "            dna_sequence = mutate_via_markov_chain(dna_sequence,mc,mutation_site)\n",
    "        else:\n",
    "            threshold = random.random()\n",
    "            ## Deletions are more likely than insertions (Zhang,2003)\n",
    "            sequenceLength = min(random.randint(1,51),round(0.05*dna_length))\n",
    "            if(threshold < 0.7):                \n",
    "                dna_sequence = mutate_deletion(dna_sequence,sequenceLength,mutation_site)\n",
    "            else:\n",
    "                ## Tandem mutations are more likely\n",
    "                threshold = random.random()\n",
    "                if(threshold<0.75):\n",
    "                    dna_sequence = mutate_tandem_insertion(dna_sequence,sequenceLength,random.randint(1,5),mutation_site)                    \n",
    "                else:\n",
    "                    dna_sequence = mutate_insertion(dna_sequence,get_gc_count(dna_sequence)/float(len(dna_sequence)),sequenceLength,mutation_site)      \n",
    "    return dna_sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_ohnologs_mutations_one(df_ohnologs_to_mutate,mutation_rate):    \n",
    "    new_sequence_1 = []\n",
    "    new_sequence_1_len = []\n",
    "    new_sequence_1_gc = []\n",
    "    new_sequence_2 = []\n",
    "    new_sequence_2_len = []\n",
    "    new_sequence_2_gc = []\n",
    "    mutated_sequence_nr = []\n",
    "    \n",
    "    for index, row in df_ohnologs_to_mutate.iterrows():        \n",
    "        threshold = random.random()\n",
    "        if(threshold<0.5):              \n",
    "            mutated_sequence = generate_mutated_sequence(row[\"Sequence-1\"],mutation_rate)\n",
    "            new_sequence_1.append(mutated_sequence)\n",
    "            new_sequence_1_len.append(len(mutated_sequence))\n",
    "            new_sequence_1_gc.append(get_gc_count(mutated_sequence)/float(len(mutated_sequence)))\n",
    "            \n",
    "            new_sequence_2.append(row['Sequence-2'])\n",
    "            new_sequence_2_len.append(row['Sequence-2 Length'])\n",
    "            new_sequence_2_gc.append(row['Sequence-2 GC'])   \n",
    "            mutated_sequence_nr.append(1)\n",
    "        else:\n",
    "            new_sequence_1.append(row['Sequence-1'])\n",
    "            new_sequence_1_len.append(row['Sequence-1 Length'])\n",
    "            new_sequence_1_gc.append(row['Sequence-1 GC'])\n",
    "            \n",
    "            mutated_sequence = generate_mutated_sequence(row[\"Sequence-2\"],mutation_rate)\n",
    "            new_sequence_2.append(mutated_sequence)\n",
    "            new_sequence_2_len.append(len(mutated_sequence))\n",
    "            new_sequence_2_gc.append(get_gc_count(mutated_sequence)/float(len(mutated_sequence)))\n",
    "            mutated_sequence_nr.append(2)\n",
    "            \n",
    "    df_ohnologs_to_mutate[\"Sequence-1-Mutated\"] = new_sequence_1\n",
    "    df_ohnologs_to_mutate[\"Sequence-1 Length-Mutated\"] = new_sequence_1_len\n",
    "    df_ohnologs_to_mutate[\"Sequence-1 GC-Mutated\"] = new_sequence_1_gc\n",
    "\n",
    "    df_ohnologs_to_mutate[\"Sequence-2-Mutated\"] = new_sequence_2\n",
    "    df_ohnologs_to_mutate[\"Sequence-2 Length-Mutated\"] = new_sequence_2_len\n",
    "    df_ohnologs_to_mutate[\"Sequence-2 GC-Mutated\"] = new_sequence_2_gc\n",
    "\n",
    "    df_ohnologs_to_mutate[\"Mutated_Sequence_Nr\"] = mutated_sequence_nr    \n",
    "    return df_ohnologs_to_mutate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_ohnologs_mutations_two(df_ohnologs_to_mutate,mutation_rate):        \n",
    "    new_sequence_1 = []\n",
    "    new_sequence_1_len = []\n",
    "    new_sequence_1_gc = []\n",
    "    new_sequence_2 = []\n",
    "    new_sequence_2_len = []\n",
    "    new_sequence_2_gc = []\n",
    "    mutated_sequence_nr = []\n",
    "    \n",
    "    for index, row in df_ohnologs_to_mutate.iterrows():                        \n",
    "        mutated_sequence = generate_mutated_sequence(row[\"Sequence-1\"],mutation_rate)\n",
    "        new_sequence_1.append(mutated_sequence)\n",
    "        new_sequence_1_len.append(len(mutated_sequence))\n",
    "        new_sequence_1_gc.append(get_gc_count(mutated_sequence)/float(len(mutated_sequence)))\n",
    "\n",
    "        mutated_sequence = generate_mutated_sequence(row[\"Sequence-2\"],mutation_rate)\n",
    "        new_sequence_2.append(mutated_sequence)\n",
    "        new_sequence_2_len.append(len(mutated_sequence))\n",
    "        new_sequence_2_gc.append(get_gc_count(mutated_sequence)/float(len(mutated_sequence)))\n",
    "        \n",
    "        mutated_sequence_nr.append(0)\n",
    "                        \n",
    "    df_ohnologs_to_mutate[\"Sequence-1-Mutated\"] = new_sequence_1\n",
    "    df_ohnologs_to_mutate[\"Sequence-1 Length-Mutated\"] = new_sequence_1_len\n",
    "    df_ohnologs_to_mutate[\"Sequence-1 GC-Mutated\"] = new_sequence_1_gc\n",
    "\n",
    "    df_ohnologs_to_mutate[\"Sequence-2-Mutated\"] = new_sequence_2\n",
    "    df_ohnologs_to_mutate[\"Sequence-2 Length-Mutated\"] = new_sequence_2_len\n",
    "    df_ohnologs_to_mutate[\"Sequence-2 GC-Mutated\"] = new_sequence_2_gc\n",
    "\n",
    "    df_ohnologs_to_mutate[\"Mutated_Sequence_Nr\"] = mutated_sequence_nr        \n",
    "    \n",
    "    return df_ohnologs_to_mutate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_ohnologs_mutations(df_animal,level):\n",
    "    ## We mutatate 10% of the sample\n",
    "    df_ohnologs_to_mutate = df_animal.sample(round(len(df_animal)*0.1),random_state=42)    \n",
    "    ## We only mutate one sequence    \n",
    "    if(level == \"Low\" or level == \"Medium\"):        \n",
    "        df_ohnologs_to_mutate = get_ohnologs_mutations_one(df_ohnologs_to_mutate,mutation_rate[level])                    \n",
    "    else:\n",
    "    # We mutate both the sequences\n",
    "        df_ohnologs_to_mutate = get_ohnologs_mutations_two(df_ohnologs_to_mutate,mutation_rate[level])        \n",
    "    return df_ohnologs_to_mutate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "def standarize_df(df):\n",
    "    df_temp = pd.DataFrame(columns=['Sequence-1 Id','Sequence-2 Id','Sequence-1','Sequence-2','Sequence-1 Length','Sequence-2 Length','Sequence-1 GC','Sequence-2 GC','Is_Ohnolog'])\n",
    "    df_temp['Sequence-1 Id'] = df[\"Ohnolog-1 Id\"]\n",
    "    df_temp['Sequence-2 Id'] = df[\"Ohnolog-2 Id\"]\n",
    "    df_temp['Sequence-1-Transcript Id'] = df[\"Ohnolog-1/Transcript-ID\"]\n",
    "    df_temp['Sequence-2-Transcript Id'] = df[\"Ohnolog-2/Transcript-ID\"]\n",
    "    df_temp['Sequence-1'] = df[\"Ohnolog-1/Sequence\"]\n",
    "    df_temp['Sequence-2'] = df[\"Ohnolog-2/Sequence\"]\n",
    "    df_temp['Sequence-1 Length'] = df[\"Ohnolog-1/Sequence-Lenght\"]\n",
    "    df_temp['Sequence-2 Length'] = df[\"Ohnolog-2/Sequence-Lenght\"]\n",
    "    df_temp['Sequence-1 GC'] = df[\"GC_Percent_1\"]\n",
    "    df_temp['Sequence-2 GC'] = df[\"GC_Percent_2\"]\n",
    "    df_temp['Is_Ohnolog'] = 1\n",
    "    return df_temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutation_levels = [\"VeryLow\",\"Low\",\"Medium\",\"High\"]\n",
    "# Mutation rate per year for human. Most researched. Similar to other mammals\n",
    "mutation_rate = {}\n",
    "mutation_rate[\"VeryLow\"] = 0.01\n",
    "mutation_rate[\"Low\"] = 0.025\n",
    "mutation_rate[\"Medium\"] = 0.05\n",
    "mutation_rate[\"High\"] = 0.1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Is_Ohnolog</th>\n",
       "      <th>Is_Paralog</th>\n",
       "      <th>Sequence-1</th>\n",
       "      <th>Sequence-1 GC</th>\n",
       "      <th>Sequence-1 Id</th>\n",
       "      <th>Sequence-1 Length</th>\n",
       "      <th>Sequence-1-Transcript Id</th>\n",
       "      <th>Sequence-2</th>\n",
       "      <th>Sequence-2 GC</th>\n",
       "      <th>Sequence-2 Id</th>\n",
       "      <th>...</th>\n",
       "      <th>Nr Gap Open_Low</th>\n",
       "      <th>Evalue_Low</th>\n",
       "      <th>Bit Score_Low</th>\n",
       "      <th>Percent Identical Matches_Total</th>\n",
       "      <th>Aligment Length_Total</th>\n",
       "      <th>Nr Mismatch_Total</th>\n",
       "      <th>Nr Gap Open_Total</th>\n",
       "      <th>Evalue_Total</th>\n",
       "      <th>Bit Score_Total</th>\n",
       "      <th>Nr Hits</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>CTTTGGAAGTCCTATGAGGGACCATTTACGGTTTCCTCAGTAATTT...</td>\n",
       "      <td>0.429090</td>\n",
       "      <td>ENSG00000095464</td>\n",
       "      <td>3307</td>\n",
       "      <td>ENST00000371447</td>\n",
       "      <td>AGTATGTTTTGCAGACAAGACCCAGAGAAGTCCAGACTGGACTTGT...</td>\n",
       "      <td>0.469856</td>\n",
       "      <td>ENSG00000132915</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.828000</td>\n",
       "      <td>22.273333</td>\n",
       "      <td>96.042529</td>\n",
       "      <td>107.352941</td>\n",
       "      <td>22.941176</td>\n",
       "      <td>2.705882</td>\n",
       "      <td>4.260000</td>\n",
       "      <td>51.652941</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>CGTTTTGGCAAGGGATTAAAGTGCTCCCCCCTGTGGCAGCAGTGAC...</td>\n",
       "      <td>0.441268</td>\n",
       "      <td>ENSG00000077684</td>\n",
       "      <td>5772</td>\n",
       "      <td>ENST00000226319</td>\n",
       "      <td>ATACAATAGTGCTCCGCGCCGCCTCAGCCGCCGCCGCCGCCCAACC...</td>\n",
       "      <td>0.450953</td>\n",
       "      <td>ENSG00000102221</td>\n",
       "      <td>...</td>\n",
       "      <td>0.272727</td>\n",
       "      <td>2.195455</td>\n",
       "      <td>23.963636</td>\n",
       "      <td>95.077926</td>\n",
       "      <td>35.592593</td>\n",
       "      <td>4.925926</td>\n",
       "      <td>0.518519</td>\n",
       "      <td>5.709458</td>\n",
       "      <td>35.981481</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>AGTCAACCTCTGGAAGTAAGTCAACTCCATTCTGAAAAAGAAGAGT...</td>\n",
       "      <td>0.357972</td>\n",
       "      <td>ENSG00000109158</td>\n",
       "      <td>11973</td>\n",
       "      <td>ENST00000264318</td>\n",
       "      <td>ACATAATCTAAGACCACAAACCACCTTGTTCCACGTGAGAAGGAAA...</td>\n",
       "      <td>0.402006</td>\n",
       "      <td>ENSG00000145863</td>\n",
       "      <td>...</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>2.126667</td>\n",
       "      <td>23.900000</td>\n",
       "      <td>92.587125</td>\n",
       "      <td>114.250000</td>\n",
       "      <td>22.500000</td>\n",
       "      <td>1.250000</td>\n",
       "      <td>1.595000</td>\n",
       "      <td>78.450000</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ACCACAGAGGCGTCTGGCTAACTCATCTCCAGACCTAAGTTGGGAA...</td>\n",
       "      <td>0.572634</td>\n",
       "      <td>ENSG00000137216</td>\n",
       "      <td>3318</td>\n",
       "      <td>ENST00000259746</td>\n",
       "      <td>CAGTCTCCTGCCATGCAAAGAAGGCTCATAGAGTTGCTTTGGAAGT...</td>\n",
       "      <td>0.516923</td>\n",
       "      <td>ENSG00000196187</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.424444</td>\n",
       "      <td>22.344444</td>\n",
       "      <td>97.938150</td>\n",
       "      <td>15.200000</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.982011</td>\n",
       "      <td>25.025000</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>GGGCCTCCTCCCTGGGGTGTGAGCAGGTCGGCGCGCCACACTTCTC...</td>\n",
       "      <td>0.468779</td>\n",
       "      <td>ENSG00000113396</td>\n",
       "      <td>3219</td>\n",
       "      <td>ENST00000262462</td>\n",
       "      <td>AGTCCTGCCCGGAACCCCCGGCAACGCGCATACGACTACACCTGCT...</td>\n",
       "      <td>0.489140</td>\n",
       "      <td>ENSG00000140284</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.409714</td>\n",
       "      <td>23.271429</td>\n",
       "      <td>90.491909</td>\n",
       "      <td>42.818182</td>\n",
       "      <td>7.818182</td>\n",
       "      <td>0.454545</td>\n",
       "      <td>0.897455</td>\n",
       "      <td>33.854545</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 44 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Is_Ohnolog  Is_Paralog                                         Sequence-1  \\\n",
       "0          1         0.0  CTTTGGAAGTCCTATGAGGGACCATTTACGGTTTCCTCAGTAATTT...   \n",
       "1          1         0.0  CGTTTTGGCAAGGGATTAAAGTGCTCCCCCCTGTGGCAGCAGTGAC...   \n",
       "2          1         0.0  AGTCAACCTCTGGAAGTAAGTCAACTCCATTCTGAAAAAGAAGAGT...   \n",
       "3          1         0.0  ACCACAGAGGCGTCTGGCTAACTCATCTCCAGACCTAAGTTGGGAA...   \n",
       "4          1         0.0  GGGCCTCCTCCCTGGGGTGTGAGCAGGTCGGCGCGCCACACTTCTC...   \n",
       "\n",
       "   Sequence-1 GC    Sequence-1 Id  Sequence-1 Length Sequence-1-Transcript Id  \\\n",
       "0       0.429090  ENSG00000095464               3307          ENST00000371447   \n",
       "1       0.441268  ENSG00000077684               5772          ENST00000226319   \n",
       "2       0.357972  ENSG00000109158              11973          ENST00000264318   \n",
       "3       0.572634  ENSG00000137216               3318          ENST00000259746   \n",
       "4       0.468779  ENSG00000113396               3219          ENST00000262462   \n",
       "\n",
       "                                          Sequence-2  Sequence-2 GC  \\\n",
       "0  AGTATGTTTTGCAGACAAGACCCAGAGAAGTCCAGACTGGACTTGT...       0.469856   \n",
       "1  ATACAATAGTGCTCCGCGCCGCCTCAGCCGCCGCCGCCGCCCAACC...       0.450953   \n",
       "2  ACATAATCTAAGACCACAAACCACCTTGTTCCACGTGAGAAGGAAA...       0.402006   \n",
       "3  CAGTCTCCTGCCATGCAAAGAAGGCTCATAGAGTTGCTTTGGAAGT...       0.516923   \n",
       "4  AGTCCTGCCCGGAACCCCCGGCAACGCGCATACGACTACACCTGCT...       0.489140   \n",
       "\n",
       "     Sequence-2 Id  ...  Nr Gap Open_Low Evalue_Low Bit Score_Low  \\\n",
       "0  ENSG00000132915  ...         0.000000   4.828000     22.273333   \n",
       "1  ENSG00000102221  ...         0.272727   2.195455     23.963636   \n",
       "2  ENSG00000145863  ...         0.166667   2.126667     23.900000   \n",
       "3  ENSG00000196187  ...         0.000000   4.424444     22.344444   \n",
       "4  ENSG00000140284  ...         0.000000   1.409714     23.271429   \n",
       "\n",
       "   Percent Identical Matches_Total  Aligment Length_Total Nr Mismatch_Total  \\\n",
       "0                        96.042529             107.352941         22.941176   \n",
       "1                        95.077926              35.592593          4.925926   \n",
       "2                        92.587125             114.250000         22.500000   \n",
       "3                        97.938150              15.200000          0.750000   \n",
       "4                        90.491909              42.818182          7.818182   \n",
       "\n",
       "  Nr Gap Open_Total Evalue_Total Bit Score_Total  Nr Hits  \n",
       "0          2.705882     4.260000       51.652941       17  \n",
       "1          0.518519     5.709458       35.981481       27  \n",
       "2          1.250000     1.595000       78.450000        8  \n",
       "3          0.000000     3.982011       25.025000       20  \n",
       "4          0.454545     0.897455       33.854545       11  \n",
       "\n",
       "[5 rows x 44 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_dict = get_df_dict(\"Intermediate\")\n",
    "df_dict[\"Ohnologs\"].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Is_Ohnolog</th>\n",
       "      <th>Is_Paralog</th>\n",
       "      <th>Sequence-1</th>\n",
       "      <th>Sequence-1 GC</th>\n",
       "      <th>Sequence-1 Id</th>\n",
       "      <th>Sequence-1 Length</th>\n",
       "      <th>Sequence-1-Transcript Id</th>\n",
       "      <th>Sequence-2</th>\n",
       "      <th>Sequence-2 GC</th>\n",
       "      <th>Sequence-2 Id</th>\n",
       "      <th>...</th>\n",
       "      <th>Nr Gap Open_Low</th>\n",
       "      <th>Evalue_Low</th>\n",
       "      <th>Bit Score_Low</th>\n",
       "      <th>Percent Identical Matches_Total</th>\n",
       "      <th>Aligment Length_Total</th>\n",
       "      <th>Nr Mismatch_Total</th>\n",
       "      <th>Nr Gap Open_Total</th>\n",
       "      <th>Evalue_Total</th>\n",
       "      <th>Bit Score_Total</th>\n",
       "      <th>Nr Hits</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ATGGCTGAGAGCGCCTCCCCGCCCTCCTCATCTGCAGCAGCCCCAG...</td>\n",
       "      <td>0.613954</td>\n",
       "      <td>ENSG00000100346</td>\n",
       "      <td>10004</td>\n",
       "      <td>ENST00000402142</td>\n",
       "      <td>CCGCCCTCCGCCGCTGCCCCCCTTTTCGTTCGCCCTCTCGGGGCGG...</td>\n",
       "      <td>0.604694</td>\n",
       "      <td>ENSG00000006283</td>\n",
       "      <td>...</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>4.218667</td>\n",
       "      <td>25.327083</td>\n",
       "      <td>94.662741</td>\n",
       "      <td>81.870370</td>\n",
       "      <td>12.703704</td>\n",
       "      <td>1.092593</td>\n",
       "      <td>3.750000</td>\n",
       "      <td>73.051852</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CACAGGCTGAGCAGTCAGGCCCACAGCATCTGACCCCAGGCCCAGC...</td>\n",
       "      <td>0.608192</td>\n",
       "      <td>ENSG00000066056</td>\n",
       "      <td>3882</td>\n",
       "      <td>ENST00000372476</td>\n",
       "      <td>GAGCTGGAGCAGCCGCCACCGCCGCCGCCGAGGGAGCCCCGGGACG...</td>\n",
       "      <td>0.441187</td>\n",
       "      <td>ENSG00000185483</td>\n",
       "      <td>...</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>5.523571</td>\n",
       "      <td>22.464286</td>\n",
       "      <td>97.329333</td>\n",
       "      <td>14.066667</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>0.266667</td>\n",
       "      <td>5.155600</td>\n",
       "      <td>23.133333</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>GTGCTGCGGCGAGCTCCGTCCAAAAGAAAATGGGGTTTGGTGTAAA...</td>\n",
       "      <td>0.541130</td>\n",
       "      <td>ENSG00000122592</td>\n",
       "      <td>2018</td>\n",
       "      <td>ENST00000242159</td>\n",
       "      <td>CGTGAGTGGGGCGGCCAATGGGTGACTGGTGCAGATTTAACTATGT...</td>\n",
       "      <td>0.587106</td>\n",
       "      <td>ENSG00000123407</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.750000</td>\n",
       "      <td>22.033333</td>\n",
       "      <td>98.351714</td>\n",
       "      <td>13.428571</td>\n",
       "      <td>0.428571</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.500143</td>\n",
       "      <td>23.528571</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CTCATTCGGGAAATGCTAAATATTTATAGTTTGGGCTCCTGGGCCC...</td>\n",
       "      <td>0.647368</td>\n",
       "      <td>ENSG00000130812</td>\n",
       "      <td>1900</td>\n",
       "      <td>ENST00000253109</td>\n",
       "      <td>AAGTACCAAGGTCTGCGGCAGGAGGAGACCGGCTCACAGGAGCAGC...</td>\n",
       "      <td>0.561502</td>\n",
       "      <td>ENSG00000120332</td>\n",
       "      <td>...</td>\n",
       "      <td>0.222222</td>\n",
       "      <td>2.433333</td>\n",
       "      <td>22.233333</td>\n",
       "      <td>97.083300</td>\n",
       "      <td>14.700000</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>2.190200</td>\n",
       "      <td>23.260000</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CCTACACACCCCTGGATCCTCTGAAATGGCAAGGGGTAGGCATGTT...</td>\n",
       "      <td>0.580530</td>\n",
       "      <td>ENSG00000102886</td>\n",
       "      <td>1397</td>\n",
       "      <td>ENST00000406256</td>\n",
       "      <td>AAAGCCCCGGCAGTGACTGGGAGGGGAACAGGAGGAGGGACAGAGG...</td>\n",
       "      <td>0.551272</td>\n",
       "      <td>ENSG00000130055</td>\n",
       "      <td>...</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>0.705500</td>\n",
       "      <td>22.800000</td>\n",
       "      <td>95.161250</td>\n",
       "      <td>16.750000</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>0.705500</td>\n",
       "      <td>22.800000</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 44 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Is_Ohnolog  Is_Paralog                                         Sequence-1  \\\n",
       "0          0         1.0  ATGGCTGAGAGCGCCTCCCCGCCCTCCTCATCTGCAGCAGCCCCAG...   \n",
       "1          0         1.0  CACAGGCTGAGCAGTCAGGCCCACAGCATCTGACCCCAGGCCCAGC...   \n",
       "2          0         1.0  GTGCTGCGGCGAGCTCCGTCCAAAAGAAAATGGGGTTTGGTGTAAA...   \n",
       "3          0         1.0  CTCATTCGGGAAATGCTAAATATTTATAGTTTGGGCTCCTGGGCCC...   \n",
       "4          0         1.0  CCTACACACCCCTGGATCCTCTGAAATGGCAAGGGGTAGGCATGTT...   \n",
       "\n",
       "   Sequence-1 GC    Sequence-1 Id Sequence-1 Length Sequence-1-Transcript Id  \\\n",
       "0       0.613954  ENSG00000100346             10004          ENST00000402142   \n",
       "1       0.608192  ENSG00000066056              3882          ENST00000372476   \n",
       "2       0.541130  ENSG00000122592              2018          ENST00000242159   \n",
       "3       0.647368  ENSG00000130812              1900          ENST00000253109   \n",
       "4       0.580530  ENSG00000102886              1397          ENST00000406256   \n",
       "\n",
       "                                          Sequence-2  Sequence-2 GC  \\\n",
       "0  CCGCCCTCCGCCGCTGCCCCCCTTTTCGTTCGCCCTCTCGGGGCGG...       0.604694   \n",
       "1  GAGCTGGAGCAGCCGCCACCGCCGCCGCCGAGGGAGCCCCGGGACG...       0.441187   \n",
       "2  CGTGAGTGGGGCGGCCAATGGGTGACTGGTGCAGATTTAACTATGT...       0.587106   \n",
       "3  AAGTACCAAGGTCTGCGGCAGGAGGAGACCGGCTCACAGGAGCAGC...       0.561502   \n",
       "4  AAAGCCCCGGCAGTGACTGGGAGGGGAACAGGAGGAGGGACAGAGG...       0.551272   \n",
       "\n",
       "     Sequence-2 Id  ... Nr Gap Open_Low Evalue_Low Bit Score_Low  \\\n",
       "0  ENSG00000006283  ...        0.333333   4.218667     25.327083   \n",
       "1  ENSG00000185483  ...        0.142857   5.523571     22.464286   \n",
       "2  ENSG00000123407  ...        0.000000   1.750000     22.033333   \n",
       "3  ENSG00000120332  ...        0.222222   2.433333     22.233333   \n",
       "4  ENSG00000130055  ...        0.750000   0.705500     22.800000   \n",
       "\n",
       "   Percent Identical Matches_Total  Aligment Length_Total Nr Mismatch_Total  \\\n",
       "0                        94.662741              81.870370         12.703704   \n",
       "1                        97.329333              14.066667          0.400000   \n",
       "2                        98.351714              13.428571          0.428571   \n",
       "3                        97.083300              14.700000          0.400000   \n",
       "4                        95.161250              16.750000          0.750000   \n",
       "\n",
       "  Nr Gap Open_Total Evalue_Total Bit Score_Total  Nr Hits  \n",
       "0          1.092593     3.750000       73.051852       54  \n",
       "1          0.266667     5.155600       23.133333       15  \n",
       "2          0.000000     1.500143       23.528571        7  \n",
       "3          0.400000     2.190200       23.260000       10  \n",
       "4          0.750000     0.705500       22.800000        4  \n",
       "\n",
       "[5 rows x 44 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_dict[\"Paralogs\"].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "for level in levelList:    \n",
    "    df_animals_dict = get_df_dict(level)    \n",
    "    for mutation_level in mutation_levels:\n",
    "        df_mutated = get_ohnologs_mutations(df_dict[\"No-Ohnologs\"],mutation_level)\n",
    "        current_file_path = train_filepath + \"/\" + level + \"/\" + \"mutated/\"\n",
    "        df_mutated.to_pickle(current_file_path + \"no-ohnologs-complete\" + \"-\" + mutation_level + \".pkl\")  \n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
