{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from random import randint\n",
    "import random\n",
    "import os\n",
    "from matplotlib import pyplot as plt\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sb\n",
    "import pickle as pkl\n",
    "import math\n",
    "random.seed(42)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mutate_simple(dna):\n",
    "    dna_list = list(dna)\n",
    "    mutation_site = random.randint(0, len(dna_list) - 1)\n",
    "    dna_list[mutation_site] = random.choice(list('ATCG'))\n",
    "    return ''.join(dna_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def draw(discrete_probdist):\n",
    "    \"\"\"\n",
    "    Draw random value from discrete probability distribution\n",
    "    represented as a dict: P(x=value) = discrete_probdist[value].\n",
    "    \"\"\"    \n",
    "    limit = 0\n",
    "    r = random.random()\n",
    "    for value in discrete_probdist:\n",
    "        limit += discrete_probdist[value]\n",
    "        if r < limit:\n",
    "            return value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Takes into account the frecuency of transitions and transversions\n",
    "def create_markov_chain():\n",
    "    markov_chain = {}\n",
    "    for from_base in 'ATGC':\n",
    "        # Generate random transition probabilities by dividing\n",
    "        # [0,1] into four intervals of random length\n",
    "        slice_points = sorted([0] + [random.random()for i in range(3)] + [1])\n",
    "        transition_probabilities = [slice_points[i+1] - slice_points[i] for i in range(4)]\n",
    "        transition_probabilities_sorted = sorted(transition_probabilities)\n",
    "        bases_string = 'ATGC'\n",
    "        if(from_base == 'A'):\n",
    "            bases_string = bases_string.replace('G','')\n",
    "            bases_string = bases_string.replace('A','')\n",
    "            bases_string = bases_string + 'A'     \n",
    "            bases_string = bases_string + 'G'            \n",
    "        if(from_base == 'G'):\n",
    "            bases_string = bases_string.replace('A','')\n",
    "            bases_string = bases_string.replace('G','')\n",
    "            bases_string = bases_string + 'G'            \n",
    "            bases_string = bases_string + 'A'            \n",
    "        if(from_base == 'C'):\n",
    "            bases_string = bases_string.replace('T','')\n",
    "            bases_string = bases_string.replace('C','')\n",
    "            bases_string = bases_string + 'C'\n",
    "            bases_string = bases_string + 'T'\n",
    "        if(from_base == 'T'):\n",
    "            bases_string = bases_string.replace('C','')\n",
    "            bases_string = bases_string.replace('T','')\n",
    "            bases_string = bases_string + 'T'   \n",
    "            bases_string = bases_string + 'C'           \n",
    "        markov_chain[from_base] = {base: p for base, p in zip(bases_string, transition_probabilities_sorted)}\n",
    "    return markov_chain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Pointwise mutation\n",
    "def mutate_via_markov_chain(dna, markov_chain,mutation_site):       \n",
    "    dna_list = list(dna)\n",
    "    from_base = dna[mutation_site]\n",
    "    if(from_base == 'N'):\n",
    "        return dna\n",
    "    to_base = draw(markov_chain[from_base])\n",
    "    dna_list[mutation_site] = to_base\n",
    "    return ''.join(dna_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_random_sequence(length,gc_percent):\n",
    "    dna_sequence = ''\n",
    "    for i in range(0,length):\n",
    "        step = random.random()\n",
    "        next_nucleotide = ''\n",
    "        if(step < gc_percent):\n",
    "            next_nucleotide =  random.choice(list('CG'))\n",
    "        else:\n",
    "            next_nucleotide =  random.choice(list('AT'))\n",
    "        dna_sequence += next_nucleotide\n",
    "    return dna_sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_gc_count(sequence):\n",
    "    count = 0\n",
    "    for c in sequence:\n",
    "        if(c == 'G'):\n",
    "            count = count + 1\n",
    "        if(c == 'C'):\n",
    "            count = count + 1            \n",
    "    return count    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def insert_str(string, str_to_insert, index):\n",
    "    return string[:index] + str_to_insert + string[index:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mutate_insertion(dna,gc_percent,insertion_length_max,mutation_site):\n",
    "    return insert_str(dna,generate_random_sequence(randint(1, insertion_length_max),gc_percent),mutation_site)        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mutate_tandem_insertion(dna,tandem_length_max,tandem_quantity_max,mutation_site):    \n",
    "    tandem_length = randint(1, tandem_length_max)\n",
    "    final_position = mutation_site + tandem_length\n",
    "    insert_string = dna[mutation_site:final_position]\n",
    "    tandem_length = randint(1, tandem_quantity_max)\n",
    "    for i in range(1,tandem_length):\n",
    "        dna = insert_str(dna,insert_string,mutation_site) \n",
    "    return dna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mutate_deletion(dna,deletion_length_max,mutation_site):\n",
    "    deletion_length =  randint(1, deletion_length_max)\n",
    "    deletion_position = mutation_site + deletion_length\n",
    "    return dna[0:mutation_site] + dna[deletion_position:]    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Likelyhood of indels are between 16% and 25%\n",
    "def generate_mutated_sequence(dna_sequence,mutation_rate):\n",
    "    mc = create_markov_chain()\n",
    "    dna_length = len(dna_sequence)\n",
    "    nr_mutations = round(dna_length * mutation_rate)  \n",
    "    for i in range(0,nr_mutations):\n",
    "        threshold = random.random()\n",
    "        mutation_site = random.randint(0, len(dna_sequence) - 1)\n",
    "        if(threshold<0.8):\n",
    "            dna_sequence = mutate_via_markov_chain(dna_sequence,mc,mutation_site)\n",
    "        else:\n",
    "            threshold = random.random()\n",
    "            ## Deletions are more likely than insertions (Zhang,2003)\n",
    "            sequenceLength = min(random.randint(1,51),round(0.05*dna_length))\n",
    "            if(threshold < 0.7):                \n",
    "                dna_sequence = mutate_deletion(dna_sequence,sequenceLength,mutation_site)\n",
    "            else:\n",
    "                ## Tandem mutations are more likely\n",
    "                threshold = random.random()\n",
    "                if(threshold<0.75):\n",
    "                    dna_sequence = mutate_tandem_insertion(dna_sequence,sequenceLength,random.randint(1,5),mutation_site)                    \n",
    "                else:\n",
    "                    dna_sequence = mutate_insertion(dna_sequence,get_gc_count(dna_sequence)/float(len(dna_sequence)),sequenceLength,mutation_site)      \n",
    "    return dna_sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_df_dict(level):\n",
    "    df_dict = {}\n",
    "    df_dict[\"Ohnologs\"] = {}\n",
    "    df_dict[\"No-Ohnologs\"] = {}\n",
    "    df_dict[\"Paralogs\"] = {}\n",
    "    for animal in animalList:        \n",
    "        current_file_path = filepath + animal + \"/\" + level + \"/\"\n",
    "        df_dict[\"Ohnologs\"][animal] = pd.read_pickle(current_file_path + animal + \"-\" + level +\"-ohnologs.pkl\")\n",
    "        df_dict[\"No-Ohnologs\"][animal] = pd.read_pickle(current_file_path + animal + \"-\" + level +\"-no-ohnologs.pkl\")\n",
    "        df_dict[\"Paralogs\"][animal] = pd.read_pickle(current_file_path + animal + \"-\" + level +\"-paralogues.pkl\")        \n",
    "    return df_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Ohnolog-1 Id</th>\n",
       "      <th>Ohnolog-2 Id</th>\n",
       "      <th>Ohnolog-1 Symbol</th>\n",
       "      <th>Ohnolog-2 Symbol</th>\n",
       "      <th>Synteny Outgroup Support</th>\n",
       "      <th>Combine q-value(self) from all vetebrates</th>\n",
       "      <th>Combine q-value(outgroup) from all vetebrates</th>\n",
       "      <th>q-value from self comparison</th>\n",
       "      <th>Combined q-value for all outgroups</th>\n",
       "      <th>q-value for Amphioxus</th>\n",
       "      <th>...</th>\n",
       "      <th>q-value for Worm</th>\n",
       "      <th>Duplication node form Ensembl</th>\n",
       "      <th>Ohnolog-1/Sequence-Lenght</th>\n",
       "      <th>Ohnolog-1/Sequence</th>\n",
       "      <th>Ohnolog-2/Sequence-Lenght</th>\n",
       "      <th>Ohnolog-2/Sequence</th>\n",
       "      <th>Ohnolog-1/Transcript-ID</th>\n",
       "      <th>Ohnolog-2/Transcript-ID</th>\n",
       "      <th>GC_Percent_1</th>\n",
       "      <th>GC_Percent_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000095464</td>\n",
       "      <td>ENSG00000132915</td>\n",
       "      <td>PDE6C</td>\n",
       "      <td>PDE6A</td>\n",
       "      <td>4.0</td>\n",
       "      <td>8.228000e-06</td>\n",
       "      <td>0.000076</td>\n",
       "      <td>7.960000e-12</td>\n",
       "      <td>1.090000e-09</td>\n",
       "      <td>2.630000e-08</td>\n",
       "      <td>...</td>\n",
       "      <td>0.9510</td>\n",
       "      <td>Vertebrata</td>\n",
       "      <td>3307</td>\n",
       "      <td>CTTTGGAAGTCCTATGAGGGACCATTTACGGTTTCCTCAGTAATTT...</td>\n",
       "      <td>5706</td>\n",
       "      <td>AGTATGTTTTGCAGACAAGACCCAGAGAAGTCCAGACTGGACTTGT...</td>\n",
       "      <td>ENST00000371447</td>\n",
       "      <td>ENST00000508173</td>\n",
       "      <td>0.429090</td>\n",
       "      <td>0.469856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000077684</td>\n",
       "      <td>ENSG00000102221</td>\n",
       "      <td>PHF17</td>\n",
       "      <td>PHF16</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.900000e-03</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>1.070000e-02</td>\n",
       "      <td>4.790000e-07</td>\n",
       "      <td>3.370000e-01</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0789</td>\n",
       "      <td>Vertebrata</td>\n",
       "      <td>5772</td>\n",
       "      <td>CGTTTTGGCAAGGGATTAAAGTGCTCCCCCCTGTGGCAGCAGTGAC...</td>\n",
       "      <td>4934</td>\n",
       "      <td>ATACAATAGTGCTCCGCGCCGCCTCAGCCGCCGCCGCCGCCCAACC...</td>\n",
       "      <td>ENST00000226319</td>\n",
       "      <td>ENST00000614628</td>\n",
       "      <td>0.441268</td>\n",
       "      <td>0.450953</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000109158</td>\n",
       "      <td>ENSG00000145863</td>\n",
       "      <td>GABRA4</td>\n",
       "      <td>GABRA6</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.923000e-09</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>1.340000e-13</td>\n",
       "      <td>2.370000e-07</td>\n",
       "      <td>4.000000e-04</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Vertebrata</td>\n",
       "      <td>11973</td>\n",
       "      <td>AGTCAACCTCTGGAAGTAAGTCAACTCCATTCTGAAAAAGAAGAGT...</td>\n",
       "      <td>2393</td>\n",
       "      <td>ACATAATCTAAGACCACAAACCACCTTGTTCCACGTGAGAAGGAAA...</td>\n",
       "      <td>ENST00000264318</td>\n",
       "      <td>ENST00000523217</td>\n",
       "      <td>0.357972</td>\n",
       "      <td>0.402006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000130758</td>\n",
       "      <td>ENSG00000173327</td>\n",
       "      <td>MAP3K10</td>\n",
       "      <td>MAP3K11</td>\n",
       "      <td>5.0</td>\n",
       "      <td>8.785000e-07</td>\n",
       "      <td>0.004300</td>\n",
       "      <td>3.700000e-06</td>\n",
       "      <td>3.300000e-03</td>\n",
       "      <td>5.444000e-01</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0853</td>\n",
       "      <td>Vertebrata</td>\n",
       "      <td>3436</td>\n",
       "      <td>CGCGCGGCCAGGCCCTCTTAGCCCTCTGCCGTTTGGGGGGCACGGG...</td>\n",
       "      <td>5605</td>\n",
       "      <td>GAAGAAGGGAGCGGGGTCGGAGCCGTCGGGGCCAAAGGAGACGGGG...</td>\n",
       "      <td>ENST00000253055</td>\n",
       "      <td>ENST00000527304</td>\n",
       "      <td>0.684226</td>\n",
       "      <td>0.591079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000166562</td>\n",
       "      <td>ENSG00000140612</td>\n",
       "      <td>SEC11C</td>\n",
       "      <td>SEC11A</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.697000e-06</td>\n",
       "      <td>0.009500</td>\n",
       "      <td>9.420000e-07</td>\n",
       "      <td>1.140000e-02</td>\n",
       "      <td>1.457000e-01</td>\n",
       "      <td>...</td>\n",
       "      <td>0.1060</td>\n",
       "      <td>Vertebrata</td>\n",
       "      <td>2054</td>\n",
       "      <td>CGGTGGGCGGGGGCCGGCAGGTGCTCCGCAGCCGTCTGTGCCACCC...</td>\n",
       "      <td>3626</td>\n",
       "      <td>AGCGATTCTGCTGCCACAACCTCCTGAGTAGCTGGGATTACAGGCG...</td>\n",
       "      <td>ENST00000509791</td>\n",
       "      <td>ENST00000558924</td>\n",
       "      <td>0.402629</td>\n",
       "      <td>0.493657</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Ohnolog-1 Id     Ohnolog-2 Id Ohnolog-1 Symbol Ohnolog-2 Symbol  \\\n",
       "0  ENSG00000095464  ENSG00000132915            PDE6C            PDE6A   \n",
       "1  ENSG00000077684  ENSG00000102221            PHF17            PHF16   \n",
       "2  ENSG00000109158  ENSG00000145863           GABRA4           GABRA6   \n",
       "3  ENSG00000130758  ENSG00000173327          MAP3K10          MAP3K11   \n",
       "4  ENSG00000166562  ENSG00000140612           SEC11C           SEC11A   \n",
       "\n",
       "   Synteny Outgroup Support  Combine q-value(self) from all vetebrates  \\\n",
       "0                       4.0                               8.228000e-06   \n",
       "1                       5.0                               1.900000e-03   \n",
       "2                       5.0                               1.923000e-09   \n",
       "3                       5.0                               8.785000e-07   \n",
       "4                       3.0                               2.697000e-06   \n",
       "\n",
       "   Combine q-value(outgroup) from all vetebrates  \\\n",
       "0                                       0.000076   \n",
       "1                                       0.000010   \n",
       "2                                       0.000013   \n",
       "3                                       0.004300   \n",
       "4                                       0.009500   \n",
       "\n",
       "   q-value from self comparison  Combined q-value for all outgroups  \\\n",
       "0                  7.960000e-12                        1.090000e-09   \n",
       "1                  1.070000e-02                        4.790000e-07   \n",
       "2                  1.340000e-13                        2.370000e-07   \n",
       "3                  3.700000e-06                        3.300000e-03   \n",
       "4                  9.420000e-07                        1.140000e-02   \n",
       "\n",
       "   q-value for Amphioxus  ...  q-value for Worm  \\\n",
       "0           2.630000e-08  ...            0.9510   \n",
       "1           3.370000e-01  ...            0.0789   \n",
       "2           4.000000e-04  ...               NaN   \n",
       "3           5.444000e-01  ...            0.0853   \n",
       "4           1.457000e-01  ...            0.1060   \n",
       "\n",
       "   Duplication node form Ensembl  Ohnolog-1/Sequence-Lenght  \\\n",
       "0                     Vertebrata                       3307   \n",
       "1                     Vertebrata                       5772   \n",
       "2                     Vertebrata                      11973   \n",
       "3                     Vertebrata                       3436   \n",
       "4                     Vertebrata                       2054   \n",
       "\n",
       "                                  Ohnolog-1/Sequence  \\\n",
       "0  CTTTGGAAGTCCTATGAGGGACCATTTACGGTTTCCTCAGTAATTT...   \n",
       "1  CGTTTTGGCAAGGGATTAAAGTGCTCCCCCCTGTGGCAGCAGTGAC...   \n",
       "2  AGTCAACCTCTGGAAGTAAGTCAACTCCATTCTGAAAAAGAAGAGT...   \n",
       "3  CGCGCGGCCAGGCCCTCTTAGCCCTCTGCCGTTTGGGGGGCACGGG...   \n",
       "4  CGGTGGGCGGGGGCCGGCAGGTGCTCCGCAGCCGTCTGTGCCACCC...   \n",
       "\n",
       "   Ohnolog-2/Sequence-Lenght  \\\n",
       "0                       5706   \n",
       "1                       4934   \n",
       "2                       2393   \n",
       "3                       5605   \n",
       "4                       3626   \n",
       "\n",
       "                                  Ohnolog-2/Sequence  Ohnolog-1/Transcript-ID  \\\n",
       "0  AGTATGTTTTGCAGACAAGACCCAGAGAAGTCCAGACTGGACTTGT...          ENST00000371447   \n",
       "1  ATACAATAGTGCTCCGCGCCGCCTCAGCCGCCGCCGCCGCCCAACC...          ENST00000226319   \n",
       "2  ACATAATCTAAGACCACAAACCACCTTGTTCCACGTGAGAAGGAAA...          ENST00000264318   \n",
       "3  GAAGAAGGGAGCGGGGTCGGAGCCGTCGGGGCCAAAGGAGACGGGG...          ENST00000253055   \n",
       "4  AGCGATTCTGCTGCCACAACCTCCTGAGTAGCTGGGATTACAGGCG...          ENST00000509791   \n",
       "\n",
       "  Ohnolog-2/Transcript-ID  GC_Percent_1 GC_Percent_2  \n",
       "0         ENST00000508173      0.429090     0.469856  \n",
       "1         ENST00000614628      0.441268     0.450953  \n",
       "2         ENST00000523217      0.357972     0.402006  \n",
       "3         ENST00000527304      0.684226     0.591079  \n",
       "4         ENST00000558924      0.402629     0.493657  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_dict = get_df_dict(\"Strict\")\n",
    "df_dict[\"Ohnologs\"][\"Human\"].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_ohnologs_mutations_one(df_ohnologs_to_mutate,mutation_rate):    \n",
    "    new_sequence_1 = []\n",
    "    new_sequence_1_len = []\n",
    "    new_sequence_1_gc = []\n",
    "    new_sequence_2 = []\n",
    "    new_sequence_2_len = []\n",
    "    new_sequence_2_gc = []\n",
    "    mutated_sequence_nr = []\n",
    "    \n",
    "    for index, row in df_ohnologs_to_mutate.iterrows():        \n",
    "        threshold = random.random()\n",
    "        if(threshold<0.5):              \n",
    "            mutated_sequence = generate_mutated_sequence(row[\"Sequence-1\"],mutation_rate)\n",
    "            new_sequence_1.append(mutated_sequence)\n",
    "            new_sequence_1_len.append(len(mutated_sequence))\n",
    "            new_sequence_1_gc.append(get_gc_count(mutated_sequence)/float(len(mutated_sequence)))\n",
    "            \n",
    "            new_sequence_2.append(row['Sequence-2'])\n",
    "            new_sequence_2_len.append(row['Sequence-2 Length'])\n",
    "            new_sequence_2_gc.append(row['Sequence-2 GC'])   \n",
    "            mutated_sequence_nr.append(1)\n",
    "        else:\n",
    "            new_sequence_1.append(row['Sequence-1'])\n",
    "            new_sequence_1_len.append(row['Sequence-1 Length'])\n",
    "            new_sequence_1_gc.append(row['Sequence-1 GC'])\n",
    "            \n",
    "            mutated_sequence = generate_mutated_sequence(row[\"Sequence-2\"],mutation_rate)\n",
    "            new_sequence_2.append(mutated_sequence)\n",
    "            new_sequence_2_len.append(len(mutated_sequence))\n",
    "            new_sequence_2_gc.append(get_gc_count(mutated_sequence)/float(len(mutated_sequence)))\n",
    "            mutated_sequence_nr.append(2)\n",
    "            \n",
    "    df_ohnologs_to_mutate[\"Sequence-1-Mutated\"] = new_sequence_1\n",
    "    df_ohnologs_to_mutate[\"Sequence-1 Length-Mutated\"] = new_sequence_1_len\n",
    "    df_ohnologs_to_mutate[\"Sequence-1 GC-Mutated\"] = new_sequence_1_gc\n",
    "\n",
    "    df_ohnologs_to_mutate[\"Sequence-2-Mutated\"] = new_sequence_2\n",
    "    df_ohnologs_to_mutate[\"Sequence-2 Length-Mutated\"] = new_sequence_2_len\n",
    "    df_ohnologs_to_mutate[\"Sequence-2 GC-Mutated\"] = new_sequence_2_gc\n",
    "\n",
    "    df_ohnologs_to_mutate[\"Mutated_Sequence_Nr\"] = mutated_sequence_nr    \n",
    "    return df_ohnologs_to_mutate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_ohnologs_mutations_two(df_ohnologs_to_mutate,mutation_rate):        \n",
    "    new_sequence_1 = []\n",
    "    new_sequence_1_len = []\n",
    "    new_sequence_1_gc = []\n",
    "    new_sequence_2 = []\n",
    "    new_sequence_2_len = []\n",
    "    new_sequence_2_gc = []\n",
    "    mutated_sequence_nr = []\n",
    "    \n",
    "    for index, row in df_ohnologs_to_mutate.iterrows():                        \n",
    "        mutated_sequence = generate_mutated_sequence(row[\"Sequence-1\"],mutation_rate)\n",
    "        new_sequence_1.append(mutated_sequence)\n",
    "        new_sequence_1_len.append(len(mutated_sequence))\n",
    "        new_sequence_1_gc.append(get_gc_count(mutated_sequence)/float(len(mutated_sequence)))\n",
    "\n",
    "        mutated_sequence = generate_mutated_sequence(row[\"Sequence-2\"],mutation_rate)\n",
    "        new_sequence_2.append(mutated_sequence)\n",
    "        new_sequence_2_len.append(len(mutated_sequence))\n",
    "        new_sequence_2_gc.append(get_gc_count(mutated_sequence)/float(len(mutated_sequence)))\n",
    "        \n",
    "        mutated_sequence_nr.append(0)\n",
    "                        \n",
    "    df_ohnologs_to_mutate[\"Sequence-1-Mutated\"] = new_sequence_1\n",
    "    df_ohnologs_to_mutate[\"Sequence-1 Length-Mutated\"] = new_sequence_1_len\n",
    "    df_ohnologs_to_mutate[\"Sequence-1 GC-Mutated\"] = new_sequence_1_gc\n",
    "\n",
    "    df_ohnologs_to_mutate[\"Sequence-2-Mutated\"] = new_sequence_2\n",
    "    df_ohnologs_to_mutate[\"Sequence-2 Length-Mutated\"] = new_sequence_2_len\n",
    "    df_ohnologs_to_mutate[\"Sequence-2 GC-Mutated\"] = new_sequence_2_gc\n",
    "\n",
    "    df_ohnologs_to_mutate[\"Mutated_Sequence_Nr\"] = mutated_sequence_nr        \n",
    "    \n",
    "    return df_ohnologs_to_mutate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_ohnologs_mutations(df_animal,level):\n",
    "    ## We mutatate 10% of the sample\n",
    "    df_ohnologs_to_mutate = df_animal.sample(round(len(df_animal)*0.1),random_state=42)    \n",
    "    ## We only mutate one sequence    \n",
    "    if(level == \"Low\" or level == \"Medium\"):        \n",
    "        df_ohnologs_to_mutate = get_ohnologs_mutations_one(df_ohnologs_to_mutate,mutation_rate[level])                    \n",
    "    else:\n",
    "    # We mutate both the sequences\n",
    "        df_ohnologs_to_mutate = get_ohnologs_mutations_two(df_ohnologs_to_mutate,mutation_rate[level])        \n",
    "    return df_ohnologs_to_mutate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "def standarize_df(df):\n",
    "    df_temp = pd.DataFrame(columns=['Sequence-1 Id','Sequence-2 Id','Sequence-1','Sequence-2','Sequence-1 Length','Sequence-2 Length','Sequence-1 GC','Sequence-2 GC','Is_Ohnolog'])\n",
    "    df_temp['Sequence-1 Id'] = df[\"Ohnolog-1 Id\"]\n",
    "    df_temp['Sequence-2 Id'] = df[\"Ohnolog-2 Id\"]\n",
    "    df_temp['Sequence-1-Transcript Id'] = df[\"Ohnolog-1/Transcript-ID\"]\n",
    "    df_temp['Sequence-2-Transcript Id'] = df[\"Ohnolog-2/Transcript-ID\"]\n",
    "    df_temp['Sequence-1'] = df[\"Ohnolog-1/Sequence\"]\n",
    "    df_temp['Sequence-2'] = df[\"Ohnolog-2/Sequence\"]\n",
    "    df_temp['Sequence-1 Length'] = df[\"Ohnolog-1/Sequence-Lenght\"]\n",
    "    df_temp['Sequence-2 Length'] = df[\"Ohnolog-2/Sequence-Lenght\"]\n",
    "    df_temp['Sequence-1 GC'] = df[\"GC_Percent_1\"]\n",
    "    df_temp['Sequence-2 GC'] = df[\"GC_Percent_2\"]\n",
    "    df_temp['Is_Ohnolog'] = 1\n",
    "    return df_temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "filepath = \"Animals/\"\n",
    "animalList = [\"Human\",\"Pig\",\"Chicken\",\"Rat\",\"Mouse\",\"Dog\"]\n",
    "levelList = [\"Strict\",\"Relaxed\",\"Intermediate\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutation_levels = [\"VeryLow\",\"Low\",\"Medium\",\"High\"]\n",
    "# Mutation rate per year for human. Most researched. Similar to other mammals\n",
    "mutation_rate = {}\n",
    "mutation_rate[\"VeryLow\"] = 0.01\n",
    "mutation_rate[\"Low\"] = 0.025\n",
    "mutation_rate[\"Medium\"] = 0.05\n",
    "mutation_rate[\"High\"] = 0.1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "for level in levelList:    \n",
    "    df_animals_dict = get_df_dict(level)\n",
    "    for animal in animalList:\n",
    "        for mutation_level in mutation_levels:\n",
    "            df_mutated = get_ohnologs_mutations(standarize_df(df_animals_dict[\"Ohnologs\"][animal]),mutation_level)\n",
    "            current_file_path = filepath + animal + \"/\" + level + \"/\" + \"mutations/\"\n",
    "            df_mutated.to_pickle(current_file_path + animal + \"-\" + level + \"-mutated-\" + mutation_level + \"-ohnologs.pkl\")                        "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
